---
date: "`r paste0('Date: ',format(Sys.time(), '%a %d %b %Y'))`"
output: 
  word_document: 
    toc_depth: 3
    toc: no
    reference_docx: "H:/systemes/R script/Fiche_CompteRendu_LFQ_v5.docx"
---

```{r setup, include=F,warning=FALSE}
library(rmarkdown)
library(tinytex)
library(rstudioapi)
library(dplyr)
library(imputeLCMD)
library(ggrepel)
library(RColorBrewer)
library(pheatmap)
library(stringr)
library(reshape2)
library(gtools)
library(knitr)
library(tcltk)
library(writexl)
library(ggplot2)
library(scales)
library(ggalt)
library(data.table)
library(tidyr)
knitr::opts_chunk$set(eval=T, echo = F, warning = F, message = F,
                      dev =c("png","svg"),
                      dpi=500)
########### MaxQ data
#The "mqpar.xml" must be copied in the folder containing proteinGroup.txt.
#the file #runningTime.txt (found in combined/proc/ folder) must be copied in the folder containing proteinGroup.txt.

########### DIANN data
#Script done for .d DIA files from Tims-TOF Bruker
#.d name file must not contain "-".
#Spectral database and its log.txt file should be at the path indicated in the log.txt file of the analysis.
#In the spectral library, the contaminant proteins' accessions should be preceded by "Conta_" to be detected as Contaminant.
#export from DIANN must be called "report.tsv"
```

```{r DataImport, results=FALSE}
MyFile<-choose.files(default = "", caption = "Select 'report.pg_matrix.tsv' or 'proteinGroups.txt' file.",multi = FALSE) 

if(str_detect(string=MyFile,pattern="proteinGroups.txt")==TRUE){MySoft<-"MaxQ"}else{MySoft<-"DIANN"}

if(MySoft=="MaxQ"){
  if(file.exists(paste0(dirname(MyFile),"/mqpar.xml"))==FALSE){
    showDialog("Error",paste0("Add mqpar.xml in the same folder as 'proteinGroups.txt'.",dirname(MyFile),"."))
    knitr::knit_exit()
    }

  if(file.exists(paste0(dirname(MyFile),"/evidence.txt"))==FALSE){
    showDialog("Error",paste0("Add evidence.txt in the same folder as 'proteinGroups.txt'.",dirname(MyFile),"."))
    knitr::knit_exit()
  }
  
}else{
    if(file.exists(paste0(dirname(MyFile),"/report.tsv"))==FALSE){
    showDialog("Error",paste0("Add report.tsv in the same folder as 'report.pg_matrix.tsv'.",dirname(MyFile),"."))
    knitr::knit_exit()
    }
    if(file.exists(paste0(dirname(MyFile),"/report.log.txt"))==FALSE){
    showDialog("Error",paste0("Add report.log.txt in the same folder as 'report.pg_matrix.tsv'.",dirname(MyFile),"."))
    knitr::knit_exit()
    }
}
  
  
if (!file.exists(paste0(dirname(MyFile),"/svg"))){
    dir.create(file.path(dirname(MyFile), "svg"))
}
knitr::opts_chunk$set(fig.path = paste0(dirname(MyFile),"/svg/"))

df <- fread(MyFile,header = T, sep = "\t")
df<-as.data.frame(df)

if(MySoft=="MaxQ"){

  
  tt <- tktoplevel(borderwidth = 5)
  tkwm.title(tt,"Type of data")
  
  #RegEx
  tkgrid(tklabel(tt,text = "Select the type of data to analyse :"),column=5,row=1,sticky="N,E,W")
  tl1  <- tklistbox(tt,
                    height = 2,
                    selectmode = "single",
                    background = "white",
                    exportselection = 0) 
  tkgrid(tl1,column=5,row=2,sticky="N,E,W") 
  
  MyTypeData  <- c("LFQ intensity","Intensity") 
  tkinsert(tl1,"end",MyTypeData[1]) 
  tkinsert(tl1,"end",MyTypeData[2]) 
  tkselection.set(tl1, 0) 
  
  
  #OK button
  MyRegEx<-""
  OnOK  <- function() { 
    MyRegEx<<-MyTypeData[as.numeric(tkcurselection(tl1)) + 1]
    MyRegEx<<-if(MyRegEx=="LFQ intensity"){"LFQ intensity (.*)"}else{"Intensity (.*)"}
    tkdestroy(tt) 
  }
  
  OK.but  <- tkbutton(tt,
                      text = "   OK   ",
                      bg='green',
                      fg='white',
                      command = OnOK) 
  tkgrid(OK.but,column=7,row=3) 
  
  #
  tkraise(tt)
  tcl("wm", "attributes", tt, topmost=TRUE)
  tkwait.window(tt)
  
  df <- df[df$Reverse == ''| is.na(df$Reverse),]
  df <- df[df$'Only identified by site' == '' | is.na(df$'Only identified by site'),] 
  ToSearch<-if(MyRegEx=="LFQ intensity (.*)"){"LFQ intensity "}else{"Intensity."}
  x<-df[,str_detect(colnames(df),ToSearch)==TRUE]
  df<-df[,c(1:2,6:8,which(colnames(df)=="Potential contaminant"))]
  df<-cbind(df,x)
  colnames(df)[4]<-"Genes"
  colnames(df)[6]<-"Potential.contaminant"
  
  x<-colnames(df)[7:ncol(df)]
  x<-str_extract(string=x,pattern=MyRegEx,group=1)
  colnames(df)[7:ncol(df)]<-x
  
  rm(ToSearch, OK.but,tl1,tt,x,OnOK)
  
}else{
  if(length(which(colnames(df)=="Protein.Ids"))==0){
    df<-df[,c(1,1,2:ncol(df))]
    colnames(df)[2]<-"Protein.Ids"
  }
  df$Potential.contaminant<-str_detect(df$Protein.Group,"Conta_")
  df[df$Potential.contaminant==TRUE,"Potential.contaminant"]<-"+"
  df[df$Potential.contaminant==FALSE,"Potential.contaminant"]<-""
  df<-df[,c(1:5,ncol(df),6:(ncol(df)-1))]
  
  x<-colnames(df)[7:ncol(df)]
  x<-str_extract(string=x,pattern=".*\\\\(.*)\\.d$",group=1)
  colnames(df)<-c(colnames(df)[1:6],x)
  
  #Regular expression to extract sample name
  x<-colnames(df)[7:ncol(df)]
  tt <- tktoplevel(borderwidth = 5)
  tkwm.title(tt, "Regular expression for sample name")
  
  tkgrid(tklabel(tt, text = "Enter regulary expresssion extracting Sample names :"),
         column=1,row=1,
         sticky="W")
  
  MyBox1Val <<- tclVar("^.{10,14}_[^_]{0,12}_(.*)_.*_.*_.*_.*$")
  box1 <- tkentry(tt,  width = 100, bg='white', textvariable = MyBox1Val)
  tkgrid( box1,column=1,row=2)
  
  
  scr.m <- tkscrollbar(tt, repeatinterval=30,command=function(...) tkyview(tlComp,...))
  tlComp  <- tklistbox(tt,
                       height = if(length(x)<30){length(x)}else{30},
                       width = max(nchar(x))+5,
                       selectmode = "single",
                       yscrollcommand=function(...) tkset(scr.m,...),
                       exportselection = 0) 
  
  
  tkgrid(tlComp,column=1,row=4,columnspan=2,sticky="N,E,W") 
  
  for (i in (1:length(x))) { 
    tkinsert(tlComp,"end",x[i]) 
  } 
  
  tkgrid(scr.m,column=2,row=4)
  
  
  MyRegEx<-""
  OnTest  <- function() { 
    
    MyRegEx<<-tclvalue(MyBox1Val)
    for(i in 1:length(x)){
      MyNewName<-str_extract(string=x[i],pattern=MyRegEx,group=1)
      tkdelete(tlComp,0)#first
      tkinsert(tlComp,"end",MyNewName) 
    }
     
  }
  
  Test.but  <- tkbutton(tt,
                        text = "   Test regular expression   ",
                        command = OnTest) 
  tkgrid(Test.but,column=1,row=3) 
  
  
  
  
  
  OnOK  <- function() { 
    MyRegEx<<-tclvalue(MyBox1Val)
    tkdestroy(tt) 
  }
  
  OK.but  <- tkbutton(tt,
                      text = "   OK   ",
                      bg='green',
                      fg='white',
                      command = OnOK) 
  tkgrid(OK.but,column=3,row=5) 
  
  #
  tkraise(tt)
  tcl("wm", "attributes", tt, topmost=TRUE)
  tkwait.window(tt)


  
  x<-colnames(df)[7:ncol(df)]
  x<-str_extract(string=x,pattern=MyRegEx,group=1)
  colnames(df)[7:ncol(df)]<-x
  
  rm(box1,MyBox1Val,OK.but,tt,x,OnOK,i,scr.m,tlComp,OnTest,Test.but)
}

```

```{r Reorder,   results=FALSE}
#extract file names
EntireNames<-names(df)
x<-names(df)[7:ncol(df)]


#dialog box
tt <- tktoplevel(borderwidth = 5)
tkwm.title(tt,"Change Order of columns")

scr.m <- tkscrollbar(tt, repeatinterval=30,command=function(...) tkyview(tlComp,...))
tlComp  <- tklistbox(tt,
                  height = if(length(x)<30){length(x)}else{30},
                  width = max(nchar(x))+5,
                  selectmode = "single",
                  background = "white",
                  yscrollcommand=function(...) tkset(scr.m,...),
                  exportselection = 0) 


tkgrid(tlComp,column=1,row=3,columnspan=2,sticky="N,E,W") 

for (i in (1:length(x))) { 
  tkinsert(tlComp,"end",x[i]) 
} 

tkgrid(scr.m,column=3,row=3)

#Up Button
  MyUp<- function() {

    pos<-as.numeric(tkcurselection(tlComp))
    if (pos == 0){ return}else{
    text1<<-x[pos+1]
    text2<<-x[pos]
    x[pos+1]<<-text2
    x[pos]<<-text1
    tkdelete(tlComp,pos)
    tkinsert(tlComp,pos-1,text1)
    tkselection.set(tlComp,pos-1)
      }

    }

 Up.but  <- tkbutton(tt,
                    text = "    Up    ",
                    bg='grey30',
                    fg='white',
                    command = MyUp) 
tkgrid(Up.but,column=4,row=1) 

#Down Button
   MyDown<- function() { 
    
    pos<-as.numeric(tkcurselection(tlComp))
    if (pos == length(x)-1){ return}else{
    text1<<-x[pos+1]
    text2<<-x[pos+2]
    x[pos+1]<<-text2
    x[pos+2]<<-text1
    tkdelete(tlComp,pos)
    tkinsert(tlComp,pos+1,text1)
    tkselection.set(tlComp,pos+1)
    }
  } 
 
 Down.but  <- tkbutton(tt,
                    text = "   Down   ",
                    bg='grey30',
                    fg='white',
                    command = MyDown) 
 tkgrid(Down.but,column=4,row=2)
  
  
#OK : order columns
OnOK  <- function() { 
  
  df2<<-df[1:6]
  
  for(i in 1:length(x)){
    MySample<<-which(colnames(df)==x[i])
    df2<<-cbind(df2,df[,MySample])
    colnames(df2)[6+i]<<-colnames(df)[MySample]
  }

  if(file.exists(paste0(dirname(MyFile),"/report_Reordered.pg_matrix.tsv"))==FALSE){
    df3<<-df2[,-6]
    colnames(df3)[6:ncol(df3)]<<-paste0("A:\\XXXXXXXXXX_XXXX_",colnames(df3)[6:ncol(df3)],"_X_X_X_X.d")
  write.table(df3,paste0(dirname(MyFile),"/report_Reordered.pg_matrix.tsv"), 
              quote=FALSE,
              row.names = FALSE,
              sep = "\t")
  }
  df<<-df2

  tkdestroy(tt) 
}
OK.but  <- tkbutton(tt,
                    text = "   OK   ",
                    bg='green',
                    fg='white',
                    command = OnOK) 
tkgrid(OK.but,column=5,row=4) 

#
tkraise(tt)
tcl("wm", "attributes", tt, topmost=TRUE)
tkwait.window(tt)

rm(MySample,x,EntireNames,df2,OK.but,OnOK,Down.but,MyDown,Up.but,MyUp,tlComp,tt,i,scr.m)


```

```{r DataImport2,  results=FALSE}
#Find project name
  x<-dirname(MyFile)
  x<-gsub(pattern="/txt",replacement="",x)
  x<-gsub(pattern="/combined",replacement="",x)
  x<-unlist(strsplit(x, split = "/",fixed=TRUE))
  ProjectName<-x[length(x)]
  rm(x)

# Experimental design file
SampleName<-names(df)[7:ncol(df)]
if(file.exists(paste0(dirname(MyFile),"/ExperimentalDesign.txt"))==FALSE){
  phenodata<-data.frame(ShortName=SampleName,Condition=SampleName)
  rownames(phenodata) <-SampleName
  fix(phenodata)
  
  
  #Add color by condition 
  Mycondition<-levels(factor(phenodata$Condition))
  NbCond<-length(Mycondition)

  MyColor<-colorRampPalette(brewer.pal(9, "Set1"))(if(NbCond>9){NbCond}else{9})
  ListColor<-MyColor[1:NbCond]
  MyCondColor<-ListColor[match(phenodata$Condition,Mycondition)]
  phenodata$CondColor<-MyCondColor
  
  write.table(phenodata,paste0(dirname(MyFile),"/ExperimentalDesign.txt"),row.names = FALSE,sep = "\t")

  rm(Mycondition,NbCond,MyColor,ListColor,MyCondColor)

}else{
  dfSampleName<-data.frame(SampleName=SampleName,MyOrder=c(1:length(SampleName)))
  phenodata<-as.data.frame(read.table(paste0(dirname(MyFile),"/ExperimentalDesign.txt"), header = T, sep = "\t", stringsAsFactors = F))
  phenodata<-merge(x = dfSampleName, y=phenodata,by.x="SampleName",by.y="ShortName")
  phenodata<-phenodata[order(phenodata$MyOrder),]
  rownames(phenodata) <-phenodata$SampleName
  phenodata<-phenodata[,-2]
  colnames(phenodata)[1]<-"ShortName"

}


#Filter Contaminant
MyConta<-df[df$Potential.contaminant == '+' & !is.na(df$Potential.contaminant),]
write.table(MyConta,paste0(dirname(MyFile),"/report.pg_matrix_Contaminants.txt"),row.names = FALSE,sep = "\t")
dfWITHconta<-df
df<-df[df$Potential.contaminant=="",]


#select features annotations (Uniprot ID, protein name, gene name)
fdataWITHconta <- dfWITHconta[,1:6]
fdata <- df[,1:6]

#select Intensities
edataWITHconta <- dfWITHconta[,7:length(colnames(df))]
edata <- df[,7:length(colnames(df))]

#log2 transform of the data
edataWITHconta <- log2(edataWITHconta)
edataWITHconta[edataWITHconta == -Inf]<-NA
edata <- log2(edata)
edata[edata == -Inf]<-NA

#remove rows that contain only NA values
edata <- edata[rowSums(is.na(edata)) != ncol(edata), ]

#drop the unused features annotations
select_index <- match(row.names(edata),row.names(fdata))
fdata <- fdata[select_index,]
fdata<-cbind(fdata,RowNamesfdata=row.names(fdata))

rm(select_index)

```



```{r ComparisonsSetup,results=FALSE}
tt <- tktoplevel(borderwidth = 5)
tkwm.title(tt,"T-test Configuration")

#Comparisons
tkgrid(tklabel(tt,text = "Select Comparisons to test."),column=1,row=1,columnspan=2,sticky="N,E,W")

MyCond<-levels(factor(phenodata$Condition))
MyComparison<-permutations(n=length(MyCond),r=2,v=MyCond)
MyComparison<-paste(MyComparison[,1],MyComparison[,2],sep="/")

scr.m <- tkscrollbar(tt, repeatinterval=30,command=function(...) tkyview(tlComp,...))
tlComp  <- tklistbox(tt,
                  height = if(length(MyComparison)<30){length(MyComparison)}else{30},
                  width = max(nchar(MyComparison))+5,
                  selectmode = "extended",
                  background = "white",
                  yscrollcommand=function(...) tkset(scr.m,...),
                  exportselection = 0) 



tkgrid(tlComp,column=1,row=2,columnspan=2,sticky="N,E,W") 

for (i in (1:length(MyComparison))) { 
  tkinsert(tlComp,"end",MyComparison[i]) 
} 

tkgrid(scr.m,column=3,row=2)


#T-test Type : "two.sided" or "less" or "greater" and "Unpaired" or "Paired"
tkgrid(tklabel(tt,text = "Select T-test type."),column=4,row=1,sticky="N,E,W")
tl1  <- tklistbox(tt,
                  height = 6,
                  selectmode = "single",
                  background = "white",
                  exportselection = 0) 

tkgrid(tl1,column=4,row=2,sticky="N,E,W") 
MyType<-paste(rep(c("two.sided", "less", "greater"),Times=2),rep(c("Unpaired","Paired"),each=3)) 
for (i in (1:6)) { 
  tkinsert(tl1,"end",MyType[i]) 
} 
tkselection.set(tl1, 0) 
 
#P-value and Q-value threshold
tkgrid(tklabel(tt,text = "Select T-test treshold."),column=5,row=1,sticky="N,E,W")
tl2  <- tklistbox(tt,
                  height = 4,
                  selectmode = "single",
                  background = "white",
                  exportselection = 0) 
tkgrid(tl2,column=5,row=2,sticky="N,E,W") 
MyListTreshold  <- c("PValue<0.05","PValue<0.01","QValue<0.05","QValue<0.01") 
for (i in (1:4)) { 
  tkinsert(tl2,"end",MyListTreshold[i]) 
} 
tkselection.set(tl2, 0) 

#Fold Change threshold
tkgrid(tklabel(tt,text = "Enter the Fold Change threshold.\n(1 means no difference)"),column=6,row=1,sticky="N,E,W")
MySignifFC<-tclVar(1)
te<-tkentry(tt,
            textvariable = MySignifFC,
            width = 6)
tkgrid(te,column=6,row=2,sticky="N,E,W")

#OK button
OnOK  <- function() { 
  MyComparison<<-MyComparison[as.numeric(tkcurselection(tlComp)) + 1]
  
  TypeTtest<<-MyType[as.numeric(tkcurselection(tl1)) + 1]
  MyAlternative<<-str_split(TypeTtest," ")[[1]][1]
  MyPaired<<-str_split(TypeTtest," ")[[1]][2]=="Paired"
  
  MySignif<<-MyListTreshold[as.numeric(tkcurselection(tl2)) + 1]
  MySignifType<<-str_sub(MySignif,1,1)# P or Q
  MySignifVal<<-as.numeric(str_sub(MySignif,8,11))
  
  MySignifFC<<-as.numeric(tclvalue(MySignifFC))
  MySignifFC<<-log2(as.numeric(MySignifFC))
  
  MyStat<<-TRUE
  
  tkdestroy(tt) 
}
OK.but  <- tkbutton(tt,
                    text = "   OK   ",
                    bg='green',
                    fg='white',
                    command = OnOK) 
tkgrid(OK.but,column=7,row=3) 

#No Statistical analysis button
OnNoStat  <- function() { 

 MyStat<<-FALSE
  
  tkdestroy(tt) 
}
NoStat.but  <- tkbutton(tt,
                    text = "No Stat",
                    bg='darkslategray',
                    fg='white',
                    command = OnNoStat) 
tkgrid(NoStat.but,column=8,row=3) 

#
tkraise(tt)
tcl("wm", "attributes", tt, topmost=TRUE)
tkwait.window(tt)


rm(MyType,OK.but,te,tl1,tl2,tlComp,tt,i,OnOK,MyListTreshold,scr.m,NoStat.but)
```

# Project Name: `r ProjectName`    
# Objective :  
  
# 1. Experimental Design  
 
```{r Sample_characteristics_kbl}
rm(df,SampleName)
MyPhenodata<-data.frame(Sample = rownames(phenodata) ,Condition=phenodata[,2])
kable(MyPhenodata, "simple")
if(MySoft=="MaxQ"){
  
  Mylog<-read.delim(paste0(dirname(MyFile),"/mqpar.xml"),sep="\t")
  
  #Find fasta database
  MyFasta<-Mylog[str_detect(Mylog[,1],"<fastaFilePath>")==TRUE ,][1]
  MyFasta<-str_extract(string=MyFasta,
                       pattern="         <fastaFilePath>(.*)</fastaFilePath>",
                       group=1)
  MyFasta<-gsub("\\\\","/",MyFasta)
  
  #Find Maxquant version
  MyVersion<-Mylog[str_detect(Mylog[,1],"<maxQuantVersion>")==TRUE ,][1]
  MyVersion<-str_extract(string=MyVersion,
                         pattern="   <maxQuantVersion>(.*)</maxQuantVersion>",
                         group=1)
  
  #Find enzyme
  MyEnzyme<-Mylog[which(str_detect(Mylog[,1],"<enzymes>"))+1,][1]
  MyEnzyme<-str_extract(string=MyEnzyme,
                        pattern="            <string>(.*)</string>",
                        group=1)
  
  
  #Find number of missed cleavages
  MyMissCleavage<-Mylog[str_detect(Mylog[,1],"<maxMissedCleavages>")==TRUE ,][1]
  MyMissCleavage<-str_extract(string=MyMissCleavage,
                              pattern="         <maxMissedCleavages>(.*)</maxMissedCleavages>",
                              group=1)
  
  #Find variable modifications
  MyVarMPT<-(which(str_detect(Mylog[,1],"<variableModifications>"))+1):(which(str_detect(Mylog[,1],"</variableModifications>"))-1)
  MyVarMPT<-Mylog[MyVarMPT,]
  MyVarMPT<-str_extract(string=MyVarMPT,
                        pattern="            <string>(.*)</string>",
                        group=1)
  MyVarMPT<-paste(MyVarMPT,collapse=" ; ")
  
  #Find fixed modifications
  MyFixMPT<-(which(str_detect(Mylog[,1],"<fixedModifications>"))+1):(which(str_detect(Mylog[,1],"</fixedModifications>"))-1)
  MyFixMPT<-Mylog[MyFixMPT,]
  MyFixMPT<-str_extract(string=MyFixMPT, 
                        pattern="            <string>(.*)</string>",
                        group=1)
  MyFixMPT<-paste(MyFixMPT,collapse=" ; ")
  
  #Find if there was normalisation
  MyNorm<-Mylog[str_detect(Mylog[,1],"<lfqSkipNorm>")==TRUE|str_detect(Mylog[,1],"<lfqNormType>")==TRUE ,]
  MyNorm<-if(str_detect(MyNorm,"False")==TRUE|str_detect(MyNorm,"1")==TRUE){"YES"}else{"NO"}
  
  
  #Find if there was Match between run
  MyMBR<-Mylog[str_detect(Mylog[,1],"<matchBetweenRuns>")==TRUE ,]
  MyMBR<-if(str_detect(MyMBR,"True")==TRUE){"YES"}else{"NO"}
  
  #Find Mass spectrometer
  MyMS<-Mylog[str_detect(Mylog[,1],"\\.d</string>")==TRUE ,]
  MyMS<-if(length(MyMS)>0){"TIMS-TOF"}else{"Orbitrap"}
  
  #FInd time of LC
  MyEvidence<-fread(paste0(dirname(MyFile),"/evidence.txt"),select="Retention time")
  MyTpsLC<-round(max(as.numeric(MyEvidence$`Retention time`,na.rm = TRUE))/60,0)
  
  #Find number of fractions
  MyNbFraction<-(which(str_detect(Mylog[,1],"<fractions>"))+1):(which(str_detect(Mylog[,1],"</fractions>"))-1)
  MyNbFraction<-Mylog[MyNbFraction,]
  MyNbFraction<-str_extract(string=MyNbFraction, 
                        pattern="      <short>(.*)</short>",
                        group=1)
  MyNbFraction<-max(as.numeric(MyNbFraction))
  if(MyNbFraction==32767){MyNbFraction<-1}
  MyTpsLC<-paste0(MyTpsLC,"h x",MyNbFraction)
  
  
  #Find folder
    MyDir<-gsub("\\\\","\\/",MyFile)
  
  #Find time of analysis
    if (file.exists(paste0(dirname(MyFile),"/#runningTimes.txt"))==TRUE){
      MyTime<-read.delim(paste0(dirname(MyFile),"/#runningTimes.txt"),sep="\t",na.strings = "NA",header = T)
      MyTime<-gsub(",",".",MyTime[,2])
      MyTime<-sum(as.numeric(MyTime),na.rm = TRUE)
      h<-MyTime/60
      j<-h/24
      s<-j/7
      m<-j/30
      x<-data.frame(nom=c("h","days","weeks","months"),val=c(h,j,s,m))
      x<-x[which(x$val>=1),]
      x<-x[which(x$val==min(x$val)),]
      MyTime<-paste(round(x$val,1),x$nom)
    }else{
      MyTime<-""
    }
    
  #Find threads
  MyThreads<-Mylog[str_detect(Mylog[,1],"<numThreads>")==TRUE ,]
  MyThreads<-str_extract(string=MyThreads, 
                         pattern="   <numThreads>(.*)</numThreads>",
                         group=1)
  MyTime<-paste0(MyTime," (",MyThreads," threads)")

}else{
  #Find spectral data base
  Mylog<-read.delim(paste0(dirname(MyFile),"/report.log.txt"),sep="\t")
  MySpectralLib<-Mylog[str_detect(Mylog[,1],"Loading spectral library")==TRUE ,][1]
  MySpectralLib<-str_sub(MySpectralLib,33,nchar(MySpectralLib))
  MySpectralLib2<-gsub("\\\\","/",MySpectralLib)
  
  #Find if there is normalisation
  MyNorm<-Mylog[str_detect(Mylog[,1],"--no-norm")==TRUE ,]
  MyNorm<-if(length(MyNorm)>0){"NO"}else{"YES"}
  
  #Find if there is Match between runs
  MyMBR<-Mylog[str_detect(Mylog[,1],"Second pass")==TRUE ,]
  MyMBR<-if(length(MyMBR)>0){"YES"}else{"NO"}
  
  #Find fasta data base
  MyFasta<-Mylog[str_detect(Mylog[,1],".fasta")==TRUE ,][1]
  MyFasta<-str_sub(MyFasta,53,nchar(MyFasta))
  MyFasta<-gsub("\\\\","/",MyFasta)
  
  #Find Mass spectrometer
  MyMS<-Mylog[str_detect(Mylog[,1],".d")==TRUE ,]
  MyMS<-if(length(MyMS)>0){"TIMS-TOF"}else{""}
  
  #Find version of DIA-NN
  MyVersion<-colnames(Mylog)
  x<-strsplit(MyVersion, split = "..",fixed=TRUE)
  MyVersion<-x[[1]][1]
  MyVersion<-str_sub(MyVersion,8,nchar(MyVersion))
  MyVersion<-paste("DIA-NN",MyVersion)
  
  #Find time of Liquid chromatography
  MyReport<- fread(paste0(dirname(MyFile),"/report.tsv"),select="RT")
  MyTpsLC<-MyReport
  MyTpsLC<-paste0(round(max(MyTpsLC)/60,0),"h")
  
  if(file.exists(gsub(".predicted.speclib", ".log.txt", MySpectralLib))==TRUE){
    #Find variables MPTs
    Mylogfasta<-read.delim(gsub(".predicted.speclib", ".log.txt", MySpectralLib),
                         sep="\t",
                         na.strings = "NA",
                         header = T)
    MyExe<-Mylogfasta[str_detect(Mylogfasta[,1],"diann.exe")==TRUE,]
    MyExe<-strsplit(MyExe, split = " ")[[1]]
    
    MyVarMPTstart<-match("--var-mod",MyExe)+1
    if(!is.na(MyVarMPTstart)){
      x<-str_detect(MyExe[(MyVarMPTstart):length(MyExe)],"--")
      x<-first(match(TRUE,x))-2
      MyVarMPT<-MyExe[(MyVarMPTstart):(MyVarMPTstart+x)]
      MyVarMPT<-paste(MyVarMPT,collapse=" ; ")
    }else{MyVarMPT<-""}
    
    #Find fixed MPTs
    MyFixMPTstart<-match("--fixed-mod",MyExe)+1
    if(!is.na(MyFixMPTstart)){
      x<-str_detect(MyExe[(MyFixMPTstart):length(MyExe)],"--")
      x<-first(match(TRUE,x))-2
      MyFixMPT<-MyExe[(MyFixMPTstart):(MyFixMPTstart+x)]
      MyFixMPT<-paste(MyFixMPT,collapse=" ; ")
    }else{MyFixMPT<-""}
    
    if(any(MyExe=="--unimod4")){
      MyFixMPT<-if(nchar(MyFixMPT)>0){
        paste(MyFixMPT,"Carbamidomethylation (C)",sep=" ; ")
      }else{
          "Carbamidomethylation (C)"}
    }
    
    #Find enzyme
    MyEnzyme<-MyExe[(match("--cut",MyExe)+1)]
    MyEnzyme<-gsub("\\*","\\\\*",MyEnzyme)
    MyEnzyme<-gsub("\\|","\\\\|",MyEnzyme)
    
    #Find MissCleavage
    MyMissCleavage<-MyExe[(match("--missed-cleavages",MyExe)+1)]
    MyMissCleavage<-gsub("\\*","\\\\*",MyMissCleavage)
    MyMissCleavage<-gsub("\\|","\\\\|",MyMissCleavage)
  }else{
      MyVarMPT<-""
      MyFixMPT<-""
      MyEnzyme<-""
      MyMissCleavage<-""
  }
  
  #Find folder
  MyDir<-gsub("\\\\","\\/",MyFile)
  
  #Find time of analysis
  MyTime<-Mylog[str_detect(Mylog[,1],"[.*]")==TRUE ,]
  MyTime<-MyTime[length(MyTime)]
  MyTime<-str_extract(MyTime,"\\[.*\\]")
  MyTime<-str_sub(MyTime,2,nchar(MyTime)-1)
  MyTime<-as.numeric(str_split(MyTime,":")[[1]][1])
  h<-MyTime/60
  j<-h/24
  s<-j/7
  m<-j/30
  x<-data.frame(nom=c("h","days","weeks","months"),val=c(h,j,s,m))
  x<-x[which(x$val>=1),]
  x<-x[which(x$val==min(x$val)),]
  MyTime<-paste(round(x$val,1),x$nom)
  
  
  #Find number of threads
  MyThreads<-Mylog[str_detect(Mylog[,1],"--threads")==TRUE ,]
  MyThreads<-strsplit(MyThreads, split = " ")[[1]]
  MyThreads<-MyThreads[which(MyThreads=="--threads")+1]
  MyTime<-paste0(MyTime," (",MyThreads," threads)")
}
```

Mass spectrometer : **`r MyMS`**    
Time of chromatography : **`r MyTpsLC`**  
Type of fragmentation : **`r if(MyMS=="TIMS-TOF"){"CID"}else{"HCD"}`**  
Software of identification and quantification : **`r MyVersion`**  
Spectral library :   
**`r if(MySoft=="DIANN"){MySpectralLib2}`**  
Fasta file :   
**`r MyFasta`**  
Variable modifications : **`r MyVarMPT`**  
Fixed modifications : **`r MyFixMPT`**  
Enzyme : **`r MyEnzyme`**  
Miss Cleavage : **`r MyMissCleavage`**  
Normalization : **`r MyNorm`**  
Match between runs : **`r MyMBR`**  
File name :  
**`r MyDir`**  
Time of analysis : **`r MyTime`**    
  
```{r CheckMass}
rm(MySpectralLib,MyNorm,MyMBR,Mylog,MyMS,MyFasta,Mylogfasta,MyExe,MyVarMPTstart,x,MyVarMPT,MyEnzyme,MyFixMPTstart,MyFixMPT,MyThreads,MyPhenodata,MyTime,j,s,m,MyDir,MyMissCleavage)
# 
# if (MySoft=="DIANN"){
#   #######A FAIRE: ajouter noms échantillons à la place des N° fichiers sur graphiques
#   Mylog<-read.delim(paste0(dirname(MyFile),"/report.log.txt"),sep="\t")
#   MyChoice<- Mylog[str_detect(Mylog[,1],"Existing .quant files will be used"),]
#   
#   if(length(MyChoice)==0){
#   MyName<-Mylog[str_detect(Mylog[,1],"File #"),]
#   MyName<-str_extract(MyName,"#.*/.*")
#   MyName<-gsub("/.*","",MyName)
#   MyName<-as.numeric(gsub("#","",MyName))
#   
#   N_injection<-Mylog[str_detect(Mylog[,1],"Loading run"),]
#   
#   #trouve s'il y a eut Match between run
#   MyMBR<-Mylog[str_detect(Mylog[,1],"Second pass")==TRUE ,]
#   MyMBR<-if(length(MyMBR)>0){"YES"}else{"NO"}
#   
#   if(MyMBR=="YES"){
#     N_injection<-N_injection[c(1:(length(N_injection)/3),(1+(length(N_injection)*2/3)):length(N_injection))]
#   }
#   
#   N_injection<-str_split(N_injection,"_")
#   for(i in 1:length(N_injection)){
#     N_injection[[i]]<-N_injection[[i]][length(N_injection[[i]])]
#   }
#   N_injection<-as.numeric(gsub(".d","",unlist(N_injection)))
#   
#   
#   RT_window<-Mylog[str_detect(Mylog[,1],"RT window"),]
#   RT_window<-gsub(" RT window set to ","",RT_window)
#   RT_window<-str_split(RT_window,"]")
#   x<-c()
#   
#   Ion_mobility_window<-Mylog[str_detect(Mylog[,1],"Ion mobility window"),]
#   Ion_mobility_window<-gsub(" Ion mobility window set to ","",Ion_mobility_window)
#   Ion_mobility_window<-str_split(Ion_mobility_window,"]")
#   y<-c()
#   
#   Recommended_MS1_mass_accuracy_setting<-Mylog[str_detect(Mylog[,1],"Recommended MS1 mass accuracy setting"),]
#   Recommended_MS1_mass_accuracy_setting<-gsub(" Recommended MS1 mass accuracy setting: ","",Recommended_MS1_mass_accuracy_setting)
#   Recommended_MS1_mass_accuracy_setting<-str_split(Recommended_MS1_mass_accuracy_setting,"]")
#   z<-c()
#   
#   for(i in 1:length(RT_window)){
#     x<-c(x,RT_window[[i]][2])
#     y<-c(y,Ion_mobility_window[[i]][2])
#     z<-c(z,Recommended_MS1_mass_accuracy_setting[[i]][2])
#   }
#   
#   RT_window<-as.numeric(x)
#   Ion_mobility_window<-as.numeric(y)
#   Recommended_MS1_mass_accuracy_setting<-as.numeric(gsub(" ppm","",z))
#   
#   rm(x,y,z)
#   
#   
#   MyStart<-Mylog[str_detect(Mylog[,1],"File #"),]
#   MyStart<-str_split(MyStart,":")
#   MyStart<-unlist(MyStart)[seq(from=1,to=length(unlist(MyStart))-1,by=2)]
#   MyStart<-as.numeric(gsub("\\[","",MyStart))
#   
#   #######"A FAIRE : MyEnd = next file or "Cross-run analysis"
#   MyEnd<-Mylog[str_detect(Mylog[,1],"] Quantification"),]
#   MyEnd<-MyEnd[!str_detect(MyEnd,"Quantification information saved to")]
#   MyEnd<-str_split(MyEnd,":")
#   MyEnd<-unlist(MyEnd)[seq(from=1,to=length(unlist(MyEnd))-1,by=2)]
#   MyEnd<-as.numeric(gsub("\\[","",MyEnd))
#   
#   x <- data.frame(FileName=MyName,
#                   RT_window=RT_window,
#                   Ion_mobility_window=Ion_mobility_window,
#                   Recommended_MS1_mass_accuracy_setting=Recommended_MS1_mass_accuracy_setting,
#                   Injection=N_injection,
#                   Start_Time=MyStart,
#                   End_Time=MyEnd,
#                   Task_Time=(MyEnd-MyStart)/60)
#   rm(RT_window,Ion_mobility_window,Recommended_MS1_mass_accuracy_setting, MyName,N_injection,MyStart,MyEnd)
#   
#   x<-x[order(x$Start_Time),]
#   # x<-x[order(x$Injection),]
#   
#   if(MyMBR=="YES"){
#   x$Injection[seq(from=1, to = (nrow(x)/2))]<-paste(x$Injection[seq(from=1, to = (nrow(x)/2))],"1st pass")
#   x$Injection[seq(from=(nrow(x)/2)+1, to = nrow(x))]<-paste(x$Injection[seq(from=(nrow(x)/2)+1, to = nrow(x))],"2nd pass")
#   }
#   
#   MySample<-factor(x$Injection,levels=x$Injection)
#   
#   
#   g<-ggplot(x, aes(x=MySample, y=RT_window)) +
#     geom_bar(stat="identity")+
#     theme_light()+
#     theme(axis.text.x = element_text(angle = 90,hjust=1,vjust=0.25),
#           axis.title.x = element_blank(),
#           axis.title.y = element_blank())+
#     ggtitle("RT Window set to ")
#   print(g)
#   
#   g<-ggplot(x, aes(x=MySample, y=Ion_mobility_window)) +
#     geom_bar(stat="identity")+
#     theme_light()+
#     theme(axis.text.x = element_text(angle = 90,hjust=1,vjust=0.25),
#           axis.title.x = element_blank(),
#           axis.title.y = element_blank())+
#     ggtitle("Ion mobility window set to ")
#   print(g)
#   
#   g<-ggplot(x, aes(x=MySample, y=Recommended_MS1_mass_accuracy_setting)) +
#     geom_bar(stat="identity")+
#     theme_light()+
#     theme(axis.text.x = element_text(angle = 90,hjust=1,vjust=0.25),
#           axis.title.x = element_blank(),
#           axis.title.y = element_blank())+
#     ggtitle("Recommended_MS1_mass_accuracy_setting (ppm)")
#   print(g)
#   
#   g<-ggplot(x, aes(x=MySample, y=Task_Time)) +
#     geom_bar(stat="identity")+
#     theme_light()+
#     theme(axis.text.x = element_text(angle = 90,hjust=1,vjust=0.25),
#           axis.title.x = element_blank(),
#           axis.title.y = element_blank())+
#     ggtitle("Time of analysis (hours)")
#   print(g)
#   
#   rm(x,i,MySample)
#   }

# }
```
  
# 2. Exploratory analysis
## 2.1 ProfilPlots  

```{r ProfilPlots,fig.width=8}
#Profilplot with contaminants
  x<-2^edataWITHconta

  x$Prot<-row.names(x)
  x<-reshape(x,
             idvar="Prot",
             varying = colnames(x)[1:(ncol(x)-1)],
             times= colnames(x)[1:(ncol(x)-1)], 
             direction = "long",
             v.name="Intensity")
    colnames(x)[2]<-"Sample"

#annotation of contaminant proteins, BSA and Trypsin
  Myfdata<-data.frame(RowNamesfdata=row.names(fdataWITHconta),
                      ProteinGroup=if(MySoft=="MaxQ"){
                            fdataWITHconta$'Protein IDs'}else{
                            fdataWITHconta$Protein.Group},
                      Genes=fdataWITHconta$Genes,
                      Contaminant=fdataWITHconta$Potential.contaminant)
  Myfdata$Type<-"Other"
  Myfdata[Myfdata$Contaminant=="+","Type"]<-"Contaminant"
  Myfdata[str_detect(Myfdata$ProteinGroup,"P00761")==TRUE,"Type"]<-"Trypsin" 
  Myfdata[str_detect(Myfdata$ProteinGroup,"P02769")==TRUE,"Type"]<-"BSA"
  
  x<-merge(x,Myfdata,by.x="Prot",by.y="RowNamesfdata",all.x=TRUE)
  x<-x[order(x$Type,decreasing=TRUE),]
   
  if(any(str_detect(x$Type,"Trypsin"))==TRUE){
  MyTrypsin<-x[str_detect(x$Type,"Trypsin")==TRUE,]
  MyAllOther<-x[str_detect(x$Type,"Trypsin")==FALSE,]
  x<-rbind(MyAllOther,MyTrypsin)
  }
  
  level_orderX <- factor(x$Sample, level = row.names(phenodata))
  level_orderProt<-factor(x$Prot, level = unique(x$Prot))
  
  OrderType<-rev(unique(x$Type))
  TypeProt<-factor(x$Type,level=OrderType)
  
  MyColors<-OrderType
  MyColors[MyColors=="BSA"]<-"#008000"
  MyColors[MyColors=="Contaminant"]<-"black"
  MyColors[MyColors=="Other"]<-"grey"
  MyColors[MyColors=="Trypsin"]<-"#FF0080"
  
  MySize<-OrderType
  MySize[MySize!="Other"]<-1
  MySize[MySize=="Other"]<-0.5
  MySize<-as.numeric(MySize)
  

ggplot(x, aes(x =  level_orderX, y = Intensity, group = level_orderProt,size = TypeProt,color=TypeProt)) + 
  geom_line()+
  scale_size_manual(values = MySize) +
  scale_colour_manual(values = MyColors)+
  theme_light()+
  theme(axis.text.x = element_text(angle = 90,hjust=1,vjust=0.25),
        axis.title.x = element_blank())+
  ggtitle("Profilplot of Intensity per proteins (Contaminants)") 


#Profilplot without contaminants
#Filter Contaminant
x<-x[x$Type=="Other",]

#Find Top5 of proteins
MyMaxLFQ<-tapply(x$Intensity, x$Prot, function(x){max(x,na.rm=TRUE)})
MyMaxLFQ<-as.data.frame(MyMaxLFQ)
MyTOP5<-row.names(top_n(MyMaxLFQ,n=5))

ExtractFirstAccession <- function(x){
   FirstAccession<-strsplit(x, split = ";") [[1]][1]
   return(FirstAccession)
 }

MyAnnotation<-distinct(x[which(x$Prot==MyTOP5),c("Prot","ProteinGroup","Genes")])
MyAnnotation$Accession <-as.character(lapply(MyAnnotation$ProteinGroup,ExtractFirstAccession))
MyAnnotation$Gene<-as.character(lapply(MyAnnotation$Genes,ExtractFirstAccession))

for(i in 1:5){
  x[x$Prot==MyTOP5[i],"Type"]<-paste(MyAnnotation$Accession[i],MyAnnotation$Gene[i])
}

#plot
  x<-x[order(x$Type),]
   
  MyOtherrows<-x[str_detect(x$Type,"Other")==TRUE,]
  MyTOP5rows<-x[str_detect(x$Type,"Other")==FALSE,]
  x<-rbind(MyOtherrows,MyTOP5rows)


  level_orderX <- factor(x$Sample, level = row.names(phenodata))
  level_orderProt<-factor(x$Prot, level = unique(x$Prot))
  
  OrderType<-rev(unique(x$Type))
  TypeProt<-factor(x$Type,level=OrderType)
  
  MyColors<-OrderType
  for(i in 1:5){
      MyColors[MyColors==OrderType[i]]<-hue_pal()(5)[i]
  }
  MyColors[MyColors=="Other"]<-"grey"

  
  MySize<-OrderType
  MySize[MySize!="Other"]<-1
  MySize[MySize=="Other"]<-0.5
  MySize<-as.numeric(MySize)
  

ggplot(x, aes(x =  level_orderX, y = Intensity, group = level_orderProt, size = TypeProt, color=TypeProt)) + 
  geom_line() +
  scale_size_manual(values = MySize) +
  scale_colour_manual(values = MyColors) +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90,hjust=1,vjust=0.25),
        axis.title.x = element_blank()) +
  ggtitle("Profilplot of LFQ intensity per proteins (TOP5 without Contaminants)") 

rm(x,Myfdata,MyAllOther,MyAnnotation,MyMaxLFQ,MyOtherrows,MyTOP5rows,MyTrypsin,level_orderProt,level_orderX,MyColors,MySize,MyTOP5,OrderType,TypeProt,i)  
```
  
## 2.2 Number of protein and missing values per sample  
**`r nrow(MyConta)+nrow(edata)`** identified proteins.  
**`r nrow(MyConta)`** suppressed contaminants.  
**`r nrow(edata)`** remaining proteins.  

```{r NbProteins}
NbProt <- colSums(!is.na(edata))

#plot of Number of protein per sample
x <- data.frame(Sample=colnames(edata),
                NbProt=NbProt,
                Condition=phenodata$Condition,
                CondColor=phenodata$CondColor)

  Conditions<-factor(x$Condition,levels=unique(x$Condition))
  MyColor<-unique(x$CondColor)
MySample<-factor(x$Sample,levels=x$Sample)
  
ggplot(x, aes(x=MySample, y=NbProt, color=Conditions,fill=Conditions)) +
  geom_bar(stat="identity")+
  ylim(0,NA)+
  scale_colour_manual(values = MyColor)+
  scale_fill_manual(values = MyColor)+
  theme_light()+
  theme(axis.text.x = element_text(angle = 90,hjust=1,vjust=0.25),
        axis.title.x = element_blank())+
  ggtitle("Number of Protein per sample") 


#plot %missing value per sample
x <- data.frame(Sample=colnames(edata),
                Pc.missing.Value=(nrow(edata)-NbProt)/nrow(edata)*100,
                Condition=phenodata$Condition,
                CondColor=phenodata$CondColor)

ggplot(x, aes(x=MySample, y=Pc.missing.Value, color=Conditions,fill=Conditions)) +
  geom_bar(stat="identity")+
  ylim(0,NA)+
  scale_colour_manual(values = MyColor)+
  scale_fill_manual(values = MyColor)+
  theme_light()+
  theme(axis.text.x = element_text(angle = 90,hjust=1,vjust=0.25),
        axis.title.x = element_blank())+
  ggtitle("Percentage of missing values per sample") 

rm(x,Conditions,MyColor,NbProt)

```

```{r PCA, fig.keep = "none"}
#principal components analysis (PCA) sur 100%VV
#Filter
pset_full <- edata[rowSums(is.na(edata)) == 0, ]

#PCA
my.prc <- prcomp(t(pset_full) , center=T, scale=F)

#Scree plot
MyScreeplot<-function(my.prc,MyTitle){

var_explained <- my.prc$sdev^2/sum(my.prc$sdev^2)
var_explained <-round(var_explained*100,1)
var_explained <- as.data.frame(var_explained)
colnames(var_explained)[1] <- "Variability.explained"

var_explained$PC <- paste("PC",c(1:nrow(var_explained)))
var_explained$PC <- factor(var_explained$PC, levels = var_explained$PC)

var_explained$CumVar<-cumsum(my.prc$sdev^2/sum(my.prc$sdev^2))*100

ggplot(var_explained) +
  geom_bar(
       aes(x=PC,
           y=Variability.explained,
           fill=TRUE),
       stat="identity")+
  geom_line(
    aes(x=1:nrow(var_explained),y =CumVar)
    ,stat="identity")+
  geom_point(
    aes(x=1:nrow(var_explained),y =CumVar))+
  geom_hline(yintercept=80,
             color="grey",
             linetype="dashed")+
  ylim(0,100)+
  theme_light()+
  theme(axis.text.x = element_text(angle = 90,hjust=1),
        axis.title.x = element_blank())+
  ylab("Explained variability (%)")+
  theme(legend.position="none")+
  ggtitle(MyTitle) 
}

#PC1 versus PC2
MyPCA<-function(my.prc,phenodata,MyTitle){
  var_explained <- my.prc$sdev^2/sum(my.prc$sdev^2)
  var_explained <- round(var_explained*100,1)
  
  Conditions<-as.factor(phenodata$Condition)
  MyColors<-levels(Conditions)
  MyColors<-match(MyColors,phenodata$Condition)
  MyColors<-phenodata$CondColor[MyColors]

  x<-as.data.frame(my.prc$x ) 
  x$Sample<-rownames(x)
  
  ggplot(x,aes(x=PC1,y=PC2,color=Conditions)) +
    geom_point(size=4) +
    scale_colour_manual(values = MyColors)+
    scale_fill_manual(values = MyColors)+
    geom_encircle(
              aes(group = Conditions,
                fill = Conditions),
              alpha = 0.1,
              size = 2,
              show.legend = FALSE,
              na.rm = TRUE)+
    geom_text_repel(data = x, 
                    aes(label = Sample),
                    nudge_y = 2,size=3,
                    segment.size  = 0.5,
                    force=1)+
    theme_bw(base_size=10) + 
    labs(x=paste0("PC1: ",var_explained[1],"%"),
         y=paste0("PC2: ",var_explained[2],"%")) +
    theme(legend.position="right")+
    ggtitle(MyTitle)
}

MyScreeplot(my.prc,"Scree Plot Log2(Intensity) of proteins with 100% Valid Values")
MyPCA(my.prc,phenodata,"PCA Log2(Intensity) of proteins with 100% Valid Values")

rm(pset_full,p1,x)

```

```{r Imputation, include=F}
FilterPreImputation <- function(data, MinPc1Condition) {

  ValidProt<-rep(NA,nrow(data))
  
  for(i in 1:nrow(data)){
    nbVV <- tapply(as.numeric(data[i,]),
                   INDEX=phenodata[,"Condition"],
                   FUN=complete.cases)#show TRUE FALSE Valid Values for each sample per group
    nbVV <- lapply(nbVV,sum)#count number of Valid Values per group
    
    for(condition in levels(factor(phenodata[,"Condition"]))) {
      nbVV[[condition]] <- nbVV[[condition]]/table(phenodata[,"Condition"])[[condition]]*100#%Valid Values per condition.
    }
    
    ValidProt[i] <- any(nbVV >= MinPc1Condition)#return TRUE if there is at least one group with 70% of valid values.
  }
  
  return(ValidProt)
}

pset_filt_imput <- edata[suppressWarnings(FilterPreImputation(edata, MinPc1Condition = 70)),]


#Imputation of left-censored missing data.
pset_filt_imput <- impute.MinProb(pset_filt_imput,q = 0.01,tune.sigma =1)

rm(FilterPreImputation)

```


```{r PCAImputation,fig.keep = "none"}

my.prc <- prcomp(t(pset_filt_imput) , center=T, scale=F)
MyScreeplot(my.prc,"Scree Plot Log2(Intensity) of proteins with 70% Valid Values in at least 1 group")
MyPCA(my.prc,phenodata,"PCA Log2(Intensity) of proteins with 70% Valid Values in at least 1 group")

```
## 2.3 Principal Componant Analysis  
Log2(LFQ intensity) matrix were filtered before imputation (only proteins with at least 70% of valid values in at least one condition were kept), (**`r nrow(pset_filt_imput)`/`r nrow(edata)`** remaining proteins).  
Z-score were calculated after imputation.  
<br>
**Scree plot**  
Blue bars present the percentage of the variability spread on each component (PC1,PC2,...) compared to the whole variability of samples cloud. If PC1 plus PC2 are close to 80% or higher (see red line), it means that the PCA plot well represent the whole variability of all samples. In other case, it means that PCA plot represents only partially all samples. Importantly, the more there is samples, the more it's difficult to obtain 80% with PC1 and PC2. Similarly, the less there are differences between conditions, the more it's difficult to obtain 80% with PC1 and PC2.   
<br>
**PCA plot**  
On this plot, the more samples are close, the more they are similar.  

```{r PCAImputationZscore, fig.height=4.5,fig.width=7}
pset_filt_imput_Zscore <-t(apply(pset_filt_imput,1,scale))
colnames(pset_filt_imput_Zscore)<-colnames(pset_filt_imput)
rm(pset_filt_imput)

my.prc <- prcomp(t(pset_filt_imput_Zscore) , center=F, scale=F)
MyScreeplot(my.prc,"Scree Plot Z-score of proteins with 70% Valid Values in at least 1 group")
MyPCA(my.prc,phenodata,"PCA Z-score of proteins with 70% Valid Values in at least 1 group")

rm(my.prc,pset_filt_imput_Zscore)
```



## 2.4 Euclidean Distance between samples  
The more euclidean distance are small, the more samples are similar.  
```{r SamplesEuclideanDist,  fig.keep = c(1,3)}
  matDist <- as.matrix(dist(t(edata)))
  matDist[matDist == 0]<-NA 
  
  MyConditions<-data.frame(Conditions=phenodata$Condition,row.names=colnames(edata))

  MycolorCond<-phenodata[,c(2,3)]
  MycolorCond<-distinct(MycolorCond)
  x<-MycolorCond[,2]
  names(x) <- MycolorCond[,1]
  MycolorCond<-list(x)
  names(MycolorCond)<-"Conditions"
  
  hmcol<- rev(colorRampPalette(brewer.pal(9, 'GnBu'))(100))
  
#heatmap of euclidean distances
  pheatmap(matDist,  
           cluster_rows=FALSE,
           cluster_cols = FALSE,
           col=hmcol,
           annotation_col = MyConditions,
           annotation_row = MyConditions,
           annotation_colors=MycolorCond,
           na_col = "black",
           show_rownames=F,
           show_colnames=T,
           annotation_names_col = FALSE,
           annotation_names_row =  FALSE,
           main = "Heatmap of Euclidean distance between samples",
           fontsize = 8,
           fontsize_col = 8,
           border_color = NA,
           angle_col=90)

  
 
  
#Boxplot euclidean distances 
  matDist<-as.data.frame(matDist)
  matDist$Sample1<-row.names(matDist)
  matDist<-reshape(matDist,
                   idvar="Sample1",
                   varying = colnames(matDist)[1:(ncol(matDist)-1)],
                   times= colnames(matDist)[1:(ncol(matDist)-1)], 
                   direction = "long",
                   v.name="Euclidean.distance")
  matDist<-matDist[is.na(matDist$Euclidean.distance)==FALSE,]
  colnames(matDist)[2]<-"Sample2"
  matDist<-merge(x=matDist,
                       y=phenodata,
                       by.x="Sample1",
                       by.y="ShortName",
                       all.x=TRUE,
                       all.y=FALSE)
colnames(matDist)[4:5]<-paste0(colnames(matDist)[4:5],"1")

 matDist<-merge(x=matDist,
                       y=phenodata,
                       by.x="Sample2",
                       by.y="ShortName",
                       all.x=TRUE,
                       all.y=FALSE)
colnames(matDist)[6:7]<-paste0(colnames(matDist)[6:7],"2")

for(i in 1:nrow(matDist)){
  matDist$Cond1_Cond2[i]<-paste(sort(c(matDist$Condition1[i],matDist$Condition2[i])),collapse = "/")
}

MyColor<-rep("lightgray",length(levels(factor(matDist$Cond1_Cond2))))

MyNames<-boxplot(data=matDist,Euclidean.distance~Cond1_Cond2)$names
MyNames<-str_split(MyNames,pattern="/")
for(i in 1:length(MyNames)){
  if(MyNames[[i]][1]==MyNames[[i]][2]){
    #Find color of the condition
    MyColor[i]<-matDist[matDist$Condition1==MyNames[[i]][1],]$CondColor1
  }
}

matDist$Cond1_Cond2 <- as.factor(matDist$Cond1_Cond2)

ggplot(matDist, aes(x=Cond1_Cond2, y=Euclidean.distance,fill=Cond1_Cond2)) + 
  geom_boxplot(outlier.shape = NA)+
  scale_fill_manual(values=MyColor)+
  geom_jitter(shape=16, position=position_jitter(0.2),size=0.5)+
  theme_light()+
  theme(axis.text.x = element_text(angle = 90,hjust=1),
        axis.title.x = element_blank())+
  theme(legend.position="none")+
  ggtitle("Boxplot of Euclidean distance between samples") 



rm(x,hmcol,matDist,MyNames,MyColor,i)

```

## 2.5 Pearson's correlation between samples  
A Pearson's correlation near of 1 means that expression proteins of samples are very similar.  
A Pearson's correlation near of 0 means that samples are uncorrelated.  
A Pearson's correlation <0 (which is very unusual) means that samples are anti-correlated.  

```{r SamplesPearsonCorrelation, fig.keep = c(1,3)}
 # heatmap échantillons correlations
  matDist <- cor(edata,use = "pairwise.complete.obs")
  matDist[matDist == 1]<-NA 
  
      hmcol<- c(colorRampPalette(c("red","sienna1"))(250),
              colorRampPalette(c("sienna1","white"))(150),
              colorRampPalette(brewer.pal(9, 'GnBu'))(100))
  
   MyBreaks<-c(seq(from=0,to=0.4999,length.out=250),
              seq(from=0.5,to=0.7999,length.out=150),
              seq(from=0.8,to=1,length.out=100))
  
  

  pheatmap(matDist,  
           cluster_rows=FALSE,
           cluster_cols = FALSE,
           color=hmcol,
           breaks=MyBreaks,
           legend_breaks=c(1,0.90,0.8,0.5,0),
           annotation_col = MyConditions,
           annotation_row = MyConditions,
           annotation_colors=MycolorCond,
           na_col = "black",
           show_rownames=F,
           show_colnames=T,
           annotation_names_col = FALSE,
           annotation_names_row =  FALSE,
           main = "Heatmap of Pearson's correlation between samples",
           fontsize = 8,
           fontsize_col = 8,
           border_color = NA,
           angle_col=90)

  
##Boxplot correlations
  matDist<-as.data.frame(matDist)
  matDist$Sample1<-row.names(matDist)
  matDist<-reshape(matDist,
                   idvar="Sample1",
                   varying = colnames(matDist)[1:(ncol(matDist)-1)],
                   times= colnames(matDist)[1:(ncol(matDist)-1)], 
                   direction = "long",
                   v.name="Pearson")
  matDist<-matDist[is.na(matDist$Pearson)==FALSE,]
  colnames(matDist)[2]<-"Sample2"
  matDist<-merge(x=matDist,
                       y=phenodata,
                       by.x="Sample1",
                       by.y="ShortName",
                       all.x=TRUE,
                       all.y=FALSE)
colnames(matDist)[4:5]<-paste0(colnames(matDist)[4:5],"1")

 matDist<-merge(x=matDist,
                       y=phenodata,
                       by.x="Sample2",
                       by.y="ShortName",
                       all.x=TRUE,
                       all.y=FALSE)
colnames(matDist)[6:7]<-paste0(colnames(matDist)[6:7],"2")

for(i in 1:nrow(matDist)){
  matDist$Cond1_Cond2[i]<-paste(sort(c(matDist$Condition1[i],matDist$Condition2[i])),collapse = "/")
}
MyColor<-rep("lightgray",length(levels(factor(matDist$Cond1_Cond2))))

MyNames<-boxplot(data=matDist,Pearson~Cond1_Cond2)$names
MyNames<-str_split(MyNames,pattern="/")
for(i in 1:length(MyNames)){
  if(MyNames[[i]][1]==MyNames[[i]][2]){
    #Find color condition
    x<-matDist[matDist$Condition1==MyNames[[i]][1],]$CondColor1
    MyColor[i]<-x
  }
}

matDist$Cond1_Cond2 <- as.factor(matDist$Cond1_Cond2)

ggplot(matDist, aes(x=Cond1_Cond2, y=Pearson,fill=Cond1_Cond2)) + 
  geom_boxplot(outlier.shape = NA)+
  scale_fill_manual(values=MyColor)+
  geom_jitter(shape=16, position=position_jitter(0.2),size=0.5)+
  theme_light()+
  theme(axis.text.x = element_text(angle = 90,hjust=1),
        axis.title.x = element_blank())+
  theme(legend.position="none")+
  ggtitle("Boxplot of Pearson's correlation between samples") 

rm(x,hmcol,matDist,MyNames,MyColor,i,MyBreaks)
  
```

## 2.6 Heatmap log2(LFQ)

```{r HeatmapLog2Intensity}
 # Select proteins with at least 1 valid value in total
  x <- edata[rowSums(is.na(edata)) != ncol(edata), ]

 # Mean intensity
  x$MeanLFQ<-apply(x,1,function(x){mean(x,na.rm = TRUE)})

 # Sort proteins by mean intensity
  x<-x[order(x$MeanLFQ,decreasing = TRUE),]

 # Delete mean columns
  x<-x %>% select(-'MeanLFQ')
  
  hmcol<- colorRampPalette(c("green","black","red"))(100)
  
 #heatmap with sample clustering
  pheatmap(x,  
           cluster_rows=FALSE,
           cluster_cols = TRUE,
           clustering_distance_cols = "euclidean",
           clustering_method = "average",
           col=hmcol,
           annotation_col = MyConditions,
           annotation_colors = MycolorCond,
           na_col = "grey",
           show_rownames=F,
           show_colnames=T,
           annotation_names_col = FALSE,
           annotation_names_row =  FALSE,
           main = "Heatmap of Log2(Intensity)",
           fontsize = 8,
           fontsize_col = 8,
           border_color = "white",
           angle_col=90)
  
  
 #heatmap without sample clustering
    pheatmap(x,  
           cluster_rows=FALSE,
           cluster_cols = FALSE,
           clustering_distance_cols = "euclidean",
           clustering_method = "average",
           col=hmcol,
           annotation_col = MyConditions,
           annotation_colors = MycolorCond,
           na_col = "grey",
           show_rownames=F,
           show_colnames=T,
           annotation_names_col = FALSE,
           annotation_names_row =  FALSE,
           main = "Heatmap of Log2(Intensity)",
           fontsize = 8,
           fontsize_col = 8,
           border_color = "white",
           angle_col=90)

rm(x,hmcol,MyConditions,MycolorCond)
```
  
# 3. Differential expression 
  
T-test **`r if(MyStat==TRUE){TypeTtest}`** were done on proteins showing at least **`r if(MyStat==TRUE){if(MyPaired==FALSE){"3 valid values in one group and at least 70% of valid values in the other group"}else{"70% of paired valid values"}}`** using log2(LFQ intensity).  
Significant threshold is **`r if(MyStat==TRUE){paste(MySignif, if(MySignifType=="Q"){"(Benjamini-Hoechberg)"},if(MySignifFC!=0){paste0(", FC>", 2^MySignifFC," or <",round(1/(2^MySignifFC),2))})}`**.  

```{r Ttest,eval=MyStat}

#T-test
TtestApply<-function(rank,x,MyListX=MyListX,MyListY=MyListY){
  
  MyTtest<-t.test(x=as.numeric(x[rank,MyListX]),
                  y=as.numeric(x[rank,MyListY]),
                  alternative = MyAlternative,
                  paired=MyPaired, 
                  var.equal=TRUE,
                  na.action="na.omit")
  MyPvalue<-MyTtest$p.value
  if(MyPaired==FALSE){
    MyFC<-MyTtest$estimate['mean of x']-MyTtest$estimate['mean of y']
  }else{
    MyFC<-MyTtest$estimate['mean difference']#'mean of the differences'
  }
  
  res<-c(MyPvalue,MyFC)
  return(res)
}


FilterPreTtest <- function(data, Comparaison, MinPc1Condition, MinNbAllCondition) {
  
  MyListX<-which(phenodata$Condition == str_split(Comp,"/")[[1]][1])
  MyListY<-which(phenodata$Condition == str_split(Comp,"/")[[1]][2])
  
  Mydata<-data[,c(MyListX,MyListY)]

  ValidProt<-rep(NA, nrow(Mydata))
  for(i in 1:nrow(Mydata)){
    
    nbVV <- tapply(as.numeric(Mydata[i,]), INDEX=phenodata[c(MyListX,MyListY),"Condition"], FUN=complete.cases)#return TRUE FALSE Valid Values for each sample per group
    nbVV <- lapply(nbVV,sum)#nb VV per group
    
    PcVV<-nbVV
    
    for(condition in str_split(Comp,"/")[[1]]) {
      PcVV[[condition]] <- PcVV[[condition]]/table(phenodata[c(MyListX,MyListY),"Condition"])[[condition]]*100#%VV per condition
    }

    ValidProt[i] <- any(PcVV >= MinPc1Condition) & nbVV[1]>=MinNbAllCondition & nbVV[2]>=MinNbAllCondition#return TRUE there is at least 1 group with 70%VV and at least 3VV in the other
  }
  
  return(ValidProt)
}

FilterPreTtestPaired <- function(data, Comparaison, MinPc1Condition) {

  MyListX<-which(phenodata$Condition == str_split(Comp,"/")[[1]][1])
  MyListY<-which(phenodata$Condition == str_split(Comp,"/")[[1]][2])
  
  Mydata<-data[,c(MyListX,MyListY)]

  ValidProt<-rep(NA, nrow(Mydata))
  for(i in 1:nrow(Mydata)){
    
    nbVV <- tapply(as.numeric(Mydata[i,]),
                   INDEX=phenodata[c(MyListX,MyListY),"Condition"], 
                   FUN=complete.cases)#return TRUE FALSE Valid Values for each sample per group
    
    nbVV<-paste(nbVV[[1]],nbVV[[2]],sep = "_")
    nbVV<-length(nbVV[nbVV=="TRUE_TRUE"])
    PcVV<-100*nbVV/length(MyListX)#%Valid Values

    ValidProt[i] <- PcVV >= MinPc1Condition #return TRUE if there is at least 70% Valid values 
  }
  
  return(ValidProt)
}

for(Comp in MyComparison){
  #position of columns of the comparison  
  MyListX<-which(phenodata$Condition == str_split(Comp,"/")[[1]][1])
  MyListY<-which(phenodata$Condition == str_split(Comp,"/")[[1]][2])
  
  
  #Filter
  if(MyPaired==FALSE){
      DataToTest <- edata[suppressWarnings(FilterPreTtest(data=edata, Comparaison=Comp, MinPc1Condition=70, MinNbAllCondition=3)),]
  }else{
      DataToTest <- edata[suppressWarnings(FilterPreTtestPaired(data=edata, Comparaison=Comp, MinPc1Condition=70)),]
  }


  
  #T-test
  Myres<-mapply(TtestApply,1:nrow(DataToTest),MoreArgs=list(DataToTest,MyListX,MyListY))
  Myres<-as.data.frame(t(Myres))
  
  #add Qvalues
  Myres$Qvalue<-p.adjust(p=Myres[,1], method = "BH")
  
  #presentation of results
  Myres<-as.data.frame(cbind(Myres[,1],Myres[,3],Myres[,2]))
  colnames(Myres)<-paste(c("Pvalue","Qvalue","Log2"),Comp)
  Myres$RowNameDataToTest<-row.names(DataToTest)

  fdata<-merge(fdata,Myres,all.x=TRUE, all.y=FALSE,by.x="RowNamesfdata",by.y="RowNameDataToTest",sort=FALSE)
}

rm(Comp,MyListX,MyListY,DataToTest,Myres,FilterPreTtest,TtestApply)


```

```{r NbVV, eval=MyStat}

VV<-as.data.frame(row.names(edata))
for(Comp in MyComparison){
  MyListX<-which(phenodata$Condition == str_split(Comp,"/")[[1]][1])
  MyListY<-which(phenodata$Condition == str_split(Comp,"/")[[1]][2])
  
    #Counting per conditions
    x<-apply(edata[,MyListX],
                       1,
                       function(x){ length(x)-sum(is.na(x))}
                       )
    y<-apply(edata[,MyListY],
                       1,
                       function(x){ length(x)-sum(is.na(x))}
                       )

    VV$MyCombin<-paste0(x,rep("_",length(x)),y)
    colnames(VV)[ncol(VV)]<-paste0("Nb Valid values ",str_split(Comp,"/")[[1]][1],"_",str_split(Comp,"/")[[1]][2])

}

#Add to fdata
fdata<-merge(fdata,VV,by.x="RowNamesfdata",by.y="row.names(edata)",all.x=TRUE)

#reorder columns fdata
nColVV<-ncol(VV)-1
startVV<-ncol(fdata)-nColVV+1

Newfdata<-as.data.frame(fdata[,1:7])

for(i in 1:nColVV){
  Newfdata<-cbind(Newfdata,fdata[,(8+((i-1)*3)):(10+((i-1)*3))])
    
  Newfdata<-cbind(Newfdata,fdata[,(startVV+i-1)])
  colnames(Newfdata)[ncol(Newfdata)]<-colnames(fdata)[(startVV+i-1)]
  
}
fdata<-Newfdata
rm(VV,MyListX,MyListY,x,y,Comp,nColVV,startVV,Newfdata,i)


```

```{r ColumnSignificant, eval=MyStat}
SeuilPQ<-1+if(MySignifType=="P"){0}else{1}

for(i in 1:length(MyComparison)){

  Comp<-MyComparison[i]
  CompX<- str_split(Comp,"/")[[1]][1]
  CompY<- str_split(Comp,"/")[[1]][2]

  MaxX<-length(which(phenodata$Condition == CompX))
  MaxY<-length(which(phenodata$Condition == CompY))

  Signif<-function(x){
    #select columns of the comparison
    x<-x[(8+((i-1)*4)):(11+((i-1)*4))]

    MyRes<-""
    if(!is.na(x[SeuilPQ])){
        if(as.numeric(x[SeuilPQ])<MySignifVal && as.numeric(x[3])>MySignifFC){
            MyRes<-paste("Increased in",CompX)
          }else if(as.numeric(x[SeuilPQ])<MySignifVal && as.numeric(x[3])<(-MySignifFC)){
            MyRes<-paste("Decreased in",CompX)
          }

        }

    VVX<-str_split(x[4],"_")[[1]][1]
    VVY<-str_split(x[4],"_")[[1]][2]

    if(VVX==0 && VVY==MaxY){
      MyRes<-paste("Disappeared in",CompX)
    }else if(VVX==MaxX && VVY==0){
      MyRes<-paste("Appeared in",CompX)
      }

    return(MyRes)
  }


  res<-apply(fdata,1,Signif)
  fdata$NewCol<-res
  colnames(fdata)[ncol(fdata)]<-paste("Significant",Comp)
}

rm(SeuilPQ,i,Comp,CompX,CompY,MaxX,MaxY,Signif,x,res)

#reorder columns of fdata
nColS<-length(MyComparison)
startS<-ncol(fdata)-nColS+1

Newfdata<-as.data.frame(fdata[,1:7])

for(i in 1:nColS){
  Newfdata<-cbind(Newfdata,fdata[,(startS+i-1)])
  colnames(Newfdata)[ncol(Newfdata)]<-colnames(fdata)[(startS+i-1)]

  Newfdata<-cbind(Newfdata,fdata[,(8+((i-1)*4)):(11+((i-1)*4))])

}
fdata<-Newfdata

rm(nColS,startS,Newfdata,i)

```

## 3.1 Number of differential proteins  
`r if(MyStat=="YES"){if(MySignifFC!=0){"(Next table take into account FC treshold.)"}}`  
  
```{r NbSignifProt, eval=MyStat}
NbComp<-length(MyComparison)
Step1<-9+5*0:(NbComp-1) #N° Pval Columns

Step2<-c()
for (i in 1:NbComp){
  Step2<-c(Step2,Step1[i]+0:2)#N°Pval, Qval, FC columns
}

MyTable<-fdata[,Step2]

# MyTable to long
MyTable$Prot<-row.names(MyTable)
MyTable<-  reshape(MyTable,
                   idvar="Prot",
                   varying = colnames(MyTable)[1:(ncol(MyTable)-1)],
                   times= colnames(MyTable)[1:(ncol(MyTable)-1)],
                   direction = "long",
                   v.name="Value")

#Add Type
x<-str_split(MyTable$time," ")

x<-unlist(x)
MyTable$Type<-x[seq(from=1,to=length(x)-1,by=2)]
MyTable$Comp<-x[seq(from=2,to=length(x),by=2)]
MyTable<-MyTable %>% select(-'time')

MyTable<-  reshape(MyTable,
                   idvar=c("Prot","Comp"),#Names of lines
                   timevar = "Type",#Names of columns
                   direction = "wide",
                   v.name="Value")#Values

#FDR pval0.05 and 0.01
xPval5pc<-MyTable[MyTable$Value.Pvalue<0.05,]
xPval5pc<-tapply(xPval5pc$Value.Qvalue,INDEX=xPval5pc$Comp,max)
xPval5pc<-round(xPval5pc*100,1)
xPval5pc<-data.frame(Comp=names(xPval5pc),FDR=xPval5pc)

xPval1pc<-MyTable[MyTable$Value.Pvalue<0.01,]
xPval1pc<-tapply(xPval1pc$Value.Qvalue,INDEX=xPval1pc$Comp,max)
xPval1pc<-round(xPval1pc*100,1)
xPval1pc<-data.frame(Comp=names(xPval1pc),FDR=xPval1pc)

xQval5pc<-MyTable[MyTable$Value.Qvalue<0.05,]
xQval5pc<-tapply(xQval5pc$Value.Pvalue,INDEX=xQval5pc$Comp,max)
xQval5pc<-round(xQval5pc,3)
xQval5pc<-data.frame(Comp=names(xQval5pc),SeuilPval=xQval5pc)

xQval1pc<-MyTable[MyTable$Value.Qvalue<0.01,]
xQval1pc<-tapply(xQval1pc$Value.Pvalue,INDEX=xQval1pc$Comp,max)
xQval1pc<-round(xQval1pc,3)
xQval1pc<-data.frame(Comp=names(xQval1pc),SeuilPval=xQval1pc)


#Add Signif
MyTable$SignifPval5pcFC<-(MyTable[,"Value.Pvalue"]<0.05
                          & (MyTable[,"Value.Log2"]<(-MySignifFC )
                          | MyTable[,"Value.Log2"]>MySignifFC))
MyTable$SignifPval1pcFC<-(MyTable[,"Value.Pvalue"]<0.01
                          & (MyTable[,"Value.Log2"]<(-MySignifFC )
                          | MyTable[,"Value.Log2"]>MySignifFC))
MyTable$SignifQval5pcFC<-(MyTable[,"Value.Qvalue"]<0.05
                          & (MyTable[,"Value.Log2"]<(-MySignifFC )
                          | MyTable[,"Value.Log2"]>MySignifFC))
MyTable$SignifQval1pcFC<-(MyTable[,"Value.Qvalue"]<0.01
                          & (MyTable[,"Value.Log2"]<(-MySignifFC )
                          | MyTable[,"Value.Log2"]>MySignifFC))

MyTable$Pval<-!is.na(MyTable[,"Value.Pvalue"])

#delete unused columns
MyTable<-MyTable %>% select(-c('Value.Pvalue','Value.Qvalue','Value.Log2'))


#Counting
Counting<-matrix(MyComparison,nrow=length(MyComparison),ncol=1)

  #Pval<0.05
x<-as.data.frame(table(MyTable$Comp,MyTable$SignifPval5pcFC))
x<-x[x$Var2==TRUE,]
x<-x[,c(1,3)]
colnames(x)[2]<-"Pvalue<0.05"
Counting<-merge(Counting,x,by.y="Var1",by.x="V1",all.x=TRUE)

  #Pval<0.01
x<-as.data.frame(table(MyTable$Comp,MyTable$SignifPval1pcFC))
x<-x[x$Var2==TRUE,]
x<-x[,c(1,3)]
colnames(x)[2]<-"Pvalue<0.01"
Counting<-merge(Counting,x,by.y="Var1",by.x="V1",all.x=TRUE)

  #Qval<0.05
x<-as.data.frame(table(MyTable$Comp,MyTable$SignifQval5pcFC))
x<-x[x$Var2==TRUE,]
x<-x[,c(1,3)]
colnames(x)[2]<-"Qvalue<0.05"
Counting<-merge(Counting,x,by.y="Var1",by.x="V1",all.x=TRUE)

  #Qval<0.01
x<-as.data.frame(table(MyTable$Comp,MyTable$SignifQval1pcFC))
x<-x[x$Var2==TRUE,]
x<-x[,c(1,3)]
colnames(x)[2]<-"Qvalue<0.01"
Counting<-merge(Counting,x,by.y="Var1",by.x="V1",all.x=TRUE)

  #Nombre de protéines testées
x<-as.data.frame(table(MyTable$Comp,MyTable$Pval))
x<-x[x$Var2==TRUE,]
x<-x[,c(1,3)]
colnames(x)[2]<-"Nb tested proteins"
Counting<-merge(Counting,x,by.y="Var1",by.x="V1",all.x=TRUE)

Counting[is.na(Counting)]<-0
colnames(Counting)[1]<-"Comparison"

#Add FDR for seuil P-val
  for(Comp in MyComparison){
    i<-which(Counting$Comparison==Comp)
    FDR5pc<-xPval5pc[which(xPval5pc$Comp==Comp),"FDR"]
    Counting[i,2]<-paste0(Counting[i,2]," (FDR=",FDR5pc,"%)")

    FDR1pc<-xPval1pc[which(xPval1pc$Comp==Comp),"FDR"]
    Counting[i,3]<-paste0(Counting[i,3]," (FDR=",FDR1pc,"%)")

    Pval5pc<-xQval5pc[which(xQval5pc$Comp==Comp),"SeuilPval"]
    Counting[i,4]<-paste0(Counting[i,4]," (Pvalue=",Pval5pc,")")

    Pval1pc<-xQval1pc[which(xQval1pc$Comp==Comp),"SeuilPval"]
    Counting[i,5]<-paste0(Counting[i,5]," (Pvalue=",Pval1pc,")")

  }
#export Counting
kable(Counting, "simple")

rm(MyTable,x,Counting, Step1, Step2, NbComp)

```

```{r NbAppearedAndDesappeared, eval=MyStat}
  NbComp<-length(MyComparison)
  Step1<-8+5*0:(NbComp-1)

  MyTable<-as.data.frame(fdata[,Step1] )

  #Counting
  Counting<-data.frame(Comparison=MyComparison,Numerator.Appeared=rep(NA,NbComp),Numerator.Disappeared=rep(NA,NbComp))

  for( i in 1:NbComp){
    MyTitle<-MyComparison[i]
    NbAppeared<-length(MyTable[str_detect(MyTable[,i],"Appeared")==TRUE,i])
    NbDisappeared<-length(MyTable[str_detect(MyTable[,i],"Disappeared")==TRUE,i])

    Counting[i,2]<-NbAppeared
    Counting[i,3]<-NbDisappeared
  }

  #export Counting
kable(Counting, "simple")

rm(MyTable,Counting, Step1, NbComp,NbAppeared,NbDisappeared,MyTitle,i)
  
```
  
## 3.2 Histogram of P-Values 
<br>  
```{r HistogrammePval,fig.width=7, eval=MyStat}
nbComp<-length(MyComparison)

#select Pvalue columns
MyColNumber<-9+((1:nbComp-1)*5)
DataToShow<-fdata[,c(2,MyColNumber)]
colnames(DataToShow)<-c("Prot",MyComparison)
DataToShow<-  reshape(DataToShow,
                   idvar="Prot",
                   varying = colnames(DataToShow)[2:ncol(DataToShow)],
                   times= colnames(DataToShow)[2:ncol(DataToShow)],
                   direction = "long",
                   v.name="Value")
DataToShow<-DataToShow[is.na(DataToShow$Value)==FALSE,]
colnames(DataToShow)[2:3]<-c("Comp", "Pvalue")


  Mycondition<-levels(factor(DataToShow$Comp))

  MyColor<-colorRampPalette(brewer.pal(8, "Set2"))(if(nbComp>8){nbComp}else{8})
  MyColor<-MyColor[1:nbComp]

#Pval graphic
p<-ggplot(DataToShow,aes(x=Pvalue,color=Comp)) +
  geom_histogram(aes(y = ..density..),alpha=0,position = "identity")+
  geom_density(alpha=0, size=1.1)+
  scale_color_manual(values=MyColor)+
  guides(color = guide_legend(title = "Comparison"))+
  theme_light()
print(p)

#-log10 Pval graphic
DataToShow$Pvalue<--log10(DataToShow$Pvalue)
colnames(DataToShow)[3]<-"Log10Pvalue"
p<-ggplot(DataToShow,aes(x=Log10Pvalue,color=Comp)) +
  geom_histogram(aes(y = ..density..),alpha=0,position = "identity")+
  geom_density(alpha=0, size=1.1)+
  scale_color_manual(values=MyColor)+
  guides(color = guide_legend(title = "Comparison"))+
  labs(x="-Log10(Pvalue)")+
  theme_light()
print(p)

rm(DataToShow,MyColNumber,MyAnnotation,nbComp,Mycondition,MyColor,mu,mu2,MyTable,MyPositionAnnotation,i,d,x)

```

    
## 3.3 Histogram of Fold change  
```{r FoldChangeHistogramme, fig.width=7, eval=MyStat}
nbComp<-length(MyComparison)
#select Fold change columns
MyColNumber<-11+((1:nbComp-1)*5) 
DataToShow<-fdata[,c(2,MyColNumber)]
colnames(DataToShow)<-c("Prot",MyComparison)
DataToShow<-  reshape(DataToShow,
                   idvar="Prot",
                   varying = colnames(DataToShow)[2:ncol(DataToShow)],
                   times= colnames(DataToShow)[2:ncol(DataToShow)],
                   direction = "long",
                   v.name="Value")
DataToShow<-DataToShow[is.na(DataToShow$Value)==FALSE,]
colnames(DataToShow)[2:3]<-c("Comp", "Log2FC")


  Mycondition<-levels(factor(DataToShow$Comp))

  MyColor<-colorRampPalette(brewer.pal(8, "Set2"))(if(nbComp>8){nbComp}else{8})
  MyColor<-MyColor[1:nbComp]

  mu<-tapply(X=DataToShow$Log2FC,INDEX=DataToShow$Comp,FUN=median)
  mu2<-tapply(X=DataToShow$Log2FC,INDEX=DataToShow$Comp,FUN=sd)
  mu<-data.frame(Comp=names(mu),Mediane=mu, Ecart.Type=mu2)

  MyAnnotation<-c()
  for(i in 1:nbComp){
    MyAnnotation<-c(MyAnnotation,mu$Comp[i],
                    paste(" Median = ",round(mu$Mediane[i],2)),
                    paste(" sd = ",round(mu$Ecart.Type[i],2)),
                    " ")
  }
    
  #detect range Fold Change
MyTable<-fdata[,c(11+((1:length(MyComparison)-1)*5))] 
MyTable<-as.data.frame(MyTable)
MyPositionAnnotation<-min(MyTable,na.rm = TRUE)  

d<-c()
for(i in 1:ncol(MyTable)){
  x<-MyTable[!is.na(MyTable[,i]),i]
  d <- max(d,max(density(x)$y))
}

  
p<-ggplot(DataToShow,aes(x=Log2FC,color=Comp)) +
  geom_density(alpha=0, size=1.1)+
  scale_color_manual(values=MyColor)+
  geom_vline(data=mu, 
             aes(xintercept=Mediane, color=Comp),
             linetype="dashed")+
  annotate("text",
              x=rep(MyPositionAnnotation,4*nbComp),
              y=seq(from= d, to= d-(d*0.05*4*nbComp), length.out= 4*nbComp),
              size=3.5,
              col = rep(MyColor,each=4),
              label = MyAnnotation,
              hjust = 0)+
  labs(y = "Density")+
  guides(color = guide_legend(title = "Comparison"))+
  theme_light()
print(p)

rm(DataToShow,MyColNumber,MyAnnotation,nbComp,Mycondition,MyColor,mu,mu2,MyTable,MyPositionAnnotation,i,d,x)

```
  
## 3.4 Volcanoplot 
Proteins in red and green are significantly differential.
<br>          
```{r Volcanoplot, fig.height=7, eval=MyStat}
n<-if(MySignifType=="P"){0}else{1}

#detect range Pvalue 
MyTable<-fdata[,c(9+((1:length(MyComparison)-1)*5))] 
MyRangePval<-c(0,max(-log10(MyTable),na.rm = TRUE))

#detect range FC
MyTable<-fdata[,c(11+((1:length(MyComparison)-1)*5))] 
MyRangeFC<-c(min(MyTable,na.rm = TRUE),max(MyTable,na.rm = TRUE))

#génère volcanoplots
for(i in 1:length(MyComparison)){
MyColNumber<-9+((i-1)*5) 

DataToShow<-fdata[is.na(fdata[,MyColNumber])==FALSE,]
DataToShow<-DataToShow[,c(MyColNumber,MyColNumber+n,MyColNumber+2,5,1)]
DataToShow$Type<-"#C4C4C4"
DataToShow[DataToShow[,2]<MySignifVal & DataToShow[,3]>MySignifFC,"Type"]<-"red"
DataToShow[DataToShow[,2]<MySignifVal & DataToShow[,3]<(-MySignifFC),"Type"]<-"limegreen"
MyY<--log10(DataToShow[,1])
MyX<-DataToShow[,3]
DataToShow$Genes<-as.character(lapply(DataToShow$Genes,ExtractFirstAccession))
DataToShow$AddGene<-""
x<-slice_min(DataToShow, order_by=DataToShow[,1], n=10)
DataToShow$AddGene[match(x$RowNamesfdata,DataToShow$RowNamesfdata)]<-"YES"
DataToShow$AddGene[DataToShow$AddGene=="YES"]<-DataToShow[DataToShow$AddGene=="YES","Genes"]

  p<-ggplot(DataToShow,aes(x=MyX,y=MyY,color=Type)) +
    geom_point(size=1) +
    scale_colour_manual(values = levels(as.factor(DataToShow$Type)))+
    geom_text_repel(data = DataToShow,
                    aes(label = AddGene),
                    segment.size  = 0.5,
                    force=1)+
    theme_bw(base_size=10) + 
    labs(x=paste0("Log2(",MyComparison[i],")"),
         y="-log10(Pvalue)") +
    xlim(MyRangeFC)+
    ylim(MyRangePval)+
    theme(legend.position="none")+
    ggtitle(paste("Volcano plot",MyComparison[i]))
  print(p)

}
rm(DataToShow,MyX,MyY,x,MyColNumber,i,MyRangeFC, MyRangePval,MyTable)

```
  
## 3.5 Heatmaps Z-score(LFQ)  
Z-scores of each protein correspond to their intensity less mean of all samples divide by their standard deviation. The more z-score is high, the more the protein is expressed in this sample compared to the others.
<br>      
```{r Heatmaps,fig.height=4, eval=MyStat}
n<-if(MySignifType=="P"){0}else{1}
for(Comp in MyComparison){
  #select columns of the comparison
  MyListX<-which(phenodata$Condition == str_split(Comp,"/")[[1]][1])
  MyListY<-which(phenodata$Condition == str_split(Comp,"/")[[1]][2])

  MyColPval<-9+((which(MyComparison==Comp)-1)*5)

  if(MyPaired==FALSE){
      #filter significative proteins
      MyListeProt<-(fdata[,
                          (MyColPval+n)]<MySignifVal
                    & (fdata[,(MyColPval+2)]<(-MySignifFC )
                       |fdata[,(MyColPval+2)]>MySignifFC))

      MyListeProt<-fdata[which(MyListeProt==TRUE),"RowNamesfdata"]
      MyListeProt<-data.frame(x1=MyListeProt,x2=MyListeProt)

      DataToShow<-cbind(edata,RowNamesedata=row.names(edata))
      DataToShow<-merge(DataToShow,MyListeProt,by.x="RowNamesedata",by.y="x1",all=FALSE)
      DataToShow<-DataToShow %>% select(-c('RowNamesedata','x2'))

      DataToShow<-DataToShow[,c(MyListX,MyListY)]

      if(nrow(MyListeProt)>1){  
        #Z-score
        MyZscore<-t(apply(DataToShow,1,scale))
        colnames(MyZscore)<-colnames(DataToShow)
  
        #heatmaps unpaired
  
        hmcol<- colorRampPalette(c("green","black","red"))(100)
        hmcol<-c("green",hmcol,"red")
  
        MyBreaks<-seq(from=-2,to=2,length.out=100)
        MinBreaks<-if(min(MyZscore,na.rm=TRUE)<(-2)){min(MyZscore,na.rm=TRUE)}else{-2.0001}
        MaxBreaks<-if(max(MyZscore,na.rm=TRUE)>2){max(MyZscore,na.rm=TRUE)}else{2.0001}
        MyBreaks<-c(MinBreaks,MyBreaks,MaxBreaks)
  
        MyConditions<-data.frame(Conditions=phenodata$Condition[c(MyListX,MyListY)],row.names=colnames(MyZscore))
  
        MycolorCond<-phenodata[c(MyListX[1],MyListY[1]),c(2,3)]
        x<-MycolorCond[,2]
        names(x) <- MycolorCond[,1]
        MycolorCond<-list(x)
        names(MycolorCond)<-"Conditions"
  
        pheatmap(MyZscore,
                 cluster_rows=T,
                 cluster_cols = T,
                 color=hmcol,
                 breaks=MyBreaks,
                 legend_breaks=c(if(MinBreaks==-2.0001){-2}else{c(round(MinBreaks,1),-2)},
                                 if(MaxBreaks==2.0001){2}else{c(2,round(MaxBreaks,1))}),
                 annotation_col = MyConditions,
                 annotation_row = NULL,
                 annotation_colors=MycolorCond,
                 na_col = "grey",
                 show_rownames=F,
                 show_colnames=T,
                 annotation_names_col = FALSE,
                 annotation_names_row =  FALSE,
                 main = paste0("Heatmap of Zscore of differential proteins ",
                              Comp,
                              " (",MySignif,if(MySignifFC!=0){paste0(", FC>",2^MySignifFC)}else{""},
                              ")"
                              ),
                 fontsize = 6,
                 fontsize_col = 8,
                 border_color = NA,
                 angle_col=90)
      }
  }else{

      #filter significative proteins

      MyListeProtIncreased<-(fdata[,(MyColPval+n)]<MySignifVal
                           & fdata[,(MyColPval+2)]>MySignifFC)

      MyListeProtDecreased<-(fdata[,(MyColPval+n)]<MySignifVal
                           & fdata[,(MyColPval+2)]<(-MySignifFC ))


      MyListeProtIncreased<-fdata[which(MyListeProtIncreased==TRUE),"RowNamesfdata"]
      MyListeProtIncreased<-data.frame(x1=MyListeProtIncreased,x2=MyListeProtIncreased)

      MyListeProtDecreased<-fdata[which(MyListeProtDecreased==TRUE),"RowNamesfdata"]
      MyListeProtDecreased<-data.frame(x1=MyListeProtDecreased,x2=MyListeProtDecreased)


      x<-length(MyListX)
      MyOrder<-rep(seq(from=1,to=x,by=1),each=2)
      MyOrder[seq(from=2,to=x*2,by=2)]<-MyOrder[seq(from=2,to=x*2,by=2)]+x


      DataToShowIncreased<-cbind(edata,RowNamesedata=row.names(edata))
      DataToShowIncreased<-merge(DataToShowIncreased,MyListeProtIncreased,by.x="RowNamesedata",by.y="x1",all=FALSE)
      DataToShowIncreased<-DataToShowIncreased %>% select(-c('RowNamesedata','x2'))
      DataToShowIncreased<-DataToShowIncreased[,c(MyListX,MyListY)]
      DataToShowIncreased<-DataToShowIncreased[,MyOrder]

      DataToShowDecreased<-cbind(edata,RowNamesedata=row.names(edata))
      DataToShowDecreased<-merge(DataToShowDecreased,MyListeProtDecreased,by.x="RowNamesedata",by.y="x1",all=FALSE)
      DataToShowDecreased<-DataToShowDecreased %>% select(-c('RowNamesedata','x2'))
      DataToShowDecreased<-DataToShowDecreased[,c(MyListX,MyListY)]
      DataToShowDecreased<-DataToShowDecreased[,MyOrder]

      #Prepare heatmap
      hmcol<- colorRampPalette(c("#FFFF80","#400080"))(100)
      hmcol<-c("#FFFF80",hmcol,"#400080")
      MyBreaks<-seq(from=-2,to=2,length.out=100)
      

      MycolorCond<-phenodata[c(MyListX[1],MyListY[1]),c(2,3)]
      x<-MycolorCond[,2]
      names(x) <- MycolorCond[,1]
      MycolorCond<-list(x)
      names(MycolorCond)<-"Conditions"

      Cond1<-str_split(Comp,"/")[[1]][1]
      Cond2<-str_split(Comp,"/")[[1]][2]
      
      if(nrow(MyListeProtIncreased)>1){ 
        #Zscore
        MyZscoreIncreased<-t(apply(DataToShowIncreased,1,scale))
        colnames(MyZscoreIncreased)<-colnames(DataToShowIncreased)
  
        MyConditions<-data.frame(Conditions=phenodata$Condition[c(MyListX,MyListY)][MyOrder],
                               row.names=colnames(MyZscoreIncreased))
        
        #heatmap Increased proteins
        MinBreaksIncreased<-if(min(MyZscoreIncreased,na.rm=TRUE)<(-2)){min(MyZscoreIncreased,na.rm=TRUE)}else{-2.0001}
        MaxBreaksIncreased<-if(max(MyZscoreIncreased,na.rm=TRUE)>2){max(MyZscoreIncreased,na.rm=TRUE)}else{2.0001}
        MyBreaksIncreased<-c(MinBreaksIncreased,MyBreaks,MaxBreaksIncreased)
  
        p<-pheatmap(MyZscoreIncreased,
                 cluster_rows=T,
                 cluster_cols = F,
                 color=hmcol,
                 legend_breaks=c(if(MinBreaksIncreased==-2.0001){-2}else{c(round(MinBreaksIncreased,1),-2)},
                                 if(MaxBreaksIncreased==2.0001){2}else{c(2,round(MaxBreaksIncreased,1))}),
                 annotation_col = MyConditions,
                 annotation_row = NULL,
                 annotation_colors=MycolorCond,
                 na_col = "white",
                 show_rownames=F,
                 show_colnames=T,
                 annotation_names_col = FALSE,
                 annotation_names_row =  FALSE,
                 main = paste0("Heatmap of Zscore of proteins increased in ",
                              Cond1," compared to ",Cond2,
                              " (",MySignif,if(MySignifFC!=0){paste0(", FC>",2^MySignifFC)}else{""},")"),
                 fontsize = 6,
                 fontsize_col = 8,
                 border_color = NA,
                 angle_col=90)
          print(p)


      }
      
      
      if(nrow(MyListeProtDecreased)>1){ 
        #Zscore
        MyZscoreDecreased<-t(apply(DataToShowDecreased,1,scale))
        colnames(MyZscoreDecreased)<-colnames(DataToShowDecreased)
        
        MyConditions<-data.frame(Conditions=phenodata$Condition[c(MyListX,MyListY)][MyOrder],
                               row.names=colnames(MyZscoreDecreased))
        
        
        #heatmap decreased proteins
        MinBreaksDecreased<-if(min(MyZscoreDecreased,na.rm=TRUE)<(-2)){min(MyZscoreDecreased,na.rm=TRUE)}else{-2.0001}
        MaxBreaksDecreased<-if(max(MyZscoreDecreased,na.rm=TRUE)>2){max(MyZscoreDecreased,na.rm=TRUE)}else{2.0001}
        MyBreaksDecreased<-c(MinBreaksDecreased,MyBreaks,MaxBreaksDecreased)

        p<-pheatmap(MyZscoreDecreased,
               cluster_rows=T,
               cluster_cols = T,
               color=hmcol,
               legend_breaks=c(if(MinBreaksDecreased==-2.0001){-2}else{c(round(MinBreaksDecreased,1),-2)},
                               if(MaxBreaksDecreased==2.0001){2}else{c(2,round(MaxBreaksDecreased,1))}),
               annotation_col = MyConditions,
               annotation_row = NULL,
               annotation_colors=MycolorCond,
               na_col = "white",
               show_rownames=F,
               show_colnames=T,
               annotation_names_col = FALSE,
               annotation_names_row =  FALSE,
               main = paste0("Heatmap of Zscore of proteins decreased in ",
                            Cond1," compared to ",Cond2,
                            " (",MySignif,if(MySignifFC!=0){paste0(", FC>",2^MySignifFC)}else{""},")"),
               fontsize = 6,
               fontsize_col = 8,
               border_color = NA,
               angle_col=90)
        print(p)
      }
  }
}

rm(Comp,MyListX,MyListY,MyColPval,hmcol,x,MycolorCond,MyConditions,MyBreaks)

if(MyPaired==FALSE){
  rm(MyListeProt,DataToShow,MyZscore,MaxBreaks,MinBreaks)
  }else{
    rm(p,MyOrder,Cond1,Cond2)
    rm(MyListeProtIncreased,DataToShowIncreased,MyZscoreIncreased,MaxBreaksIncreased,MinBreaksIncreased)
    rm(MyListeProtDecreased,DataToShowDecreased,MyZscoreDecreased,MaxBreaksDecreased,MinBreaksDecreased)
  }

```

```{r PeptideCount,results=FALSE}
if(MySoft=="DIANN"){
  # MyFile<-"H:/Users/Marjorie/TestRML_diann/BeTE230404/report.pg_matrix.tsv"
  # MyRegEx<-"^.{10,14}_[^_]{0,12}_(.*)_.*_.*_.*_.*$"
  #import peptides data
  Pepdata<-fread(paste0(dirname(MyFile),"/report.pr_matrix.tsv"),sep="\t",header=TRUE)
  Pepdata<-as.data.frame(Pepdata)
  
  
  #extract sample name
  x<-colnames(Pepdata)[11:ncol(Pepdata)]
  x<-str_extract(string=x,pattern=".*\\\\(.*)\\.d$",group=1)
  x<-str_extract(string=x,pattern=MyRegEx,group=1)
  colnames(Pepdata)[11:ncol(Pepdata)]<-x
  
  #Count peptide number
  x<-Pepdata[,c("Protein.Group","Precursor.Id")]
  x<-cbind(x,Pepdata[,11:ncol(Pepdata)])
  x<-melt(x,
          id.vars=c("Protein.Group","Precursor.Id"),
          measured.vars=colnames(Pepdata)[3:ncol(Pepdata)],
          na.rm = TRUE)
  x<-dcast(x, Protein.Group ~ variable,value.var="value",fun.aggregate=length)
  colnames(x)[2:ncol(x)]<-paste("Peptides",colnames(x)[2:ncol(x)])
  PepCount<-x
  
  #Prepare Pepdata
  Pepdata$Gene_ProteinGroup<-paste(Pepdata$Genes,Pepdata$Protein.Group)
  Pepdata<-Pepdata[,-c(1:9)]
  Pepdata<-Pepdata[,c(1,ncol(Pepdata),2:(ncol(Pepdata)-1))]
  colnames(Pepdata)[1]<-"Peptide"
  
}else{
  # MyFile<-"H:/Users/Marjorie/TestRML_diann/ErPA230331_Ubi_MaxQ/txt/proteinGroups.txt"
  #import peptides data
  Pepdata<-fread(paste0(dirname(MyFile),"/peptides.txt"),sep="\t",header=TRUE)
  Pepdata<-as.data.frame(Pepdata)
  x<-Pepdata[,c("Sequence","Protein group IDs")]
  Pepdata<-cbind(x,Pepdata[,which(str_detect(colnames(Pepdata),"Intensity "))])
  Pepdata<-separate_rows(Pepdata,'Protein group IDs',sep=";")
  x<-fread(MyFile,select=c("Protein IDs","Gene names","id"))
  x$Gene_ProteinGroup<-paste(x$`Gene names`,x$`Protein IDs`)
  x<-x[,3:4]
  x$id<-as.character(x$id)
  Pepdata<-left_join(Pepdata, x, by = join_by('Protein group IDs' == id))
  Pepdata<-Pepdata[,-2]
  Pepdata<-Pepdata[,c(1,ncol(Pepdata),2:(ncol(Pepdata)-1))]
  colnames(Pepdata)[1]<-"Peptide"
  colnames(Pepdata)[3:ncol(Pepdata)]<-str_extract(colnames(Pepdata)[3:ncol(Pepdata)],"Intensity (.*)",group=1)
  Pepdata[Pepdata==0]<-NA
  
  #Count peptide number
  x<-fread(MyFile)
  x<-as.data.frame(x)

  x<-x[,c(1,which(str_detect(colnames(x),"Peptides ")))]
  PepCount<-x
}

#Reorder PepCount

  MyOrder<-colnames(edata)
  PepCount<-as.data.frame(PepCount)
  x<-as.data.frame(PepCount[,1])
  colnames(x)<-"Protein.Group"
  for(i in 1:length(MyOrder)){
      x<-cbind(x, PepCount[,which(colnames(PepCount)==paste("Peptides",MyOrder[i]))] )
      colnames(x)[i+1]<-paste("Peptides",MyOrder[i])
  }
  PepCount<-x

#Add to fdata
if(MySoft=="MaxQ"){
  fdata<-left_join(fdata, PepCount, by = join_by('Protein IDs' == Protein.Group))
}else{
  fdata<-left_join(fdata, PepCount, by = join_by(Protein.Group == Protein.Group))
}

```

```{r ExportTable}
#Merge dfWithConta end fdata
dfWITHconta$RowNamedfWITHconta<-row.names(dfWITHconta)

  MyNcol<-ncol(dfWITHconta)
  dfWITHconta<-merge(dfWITHconta,fdata,by.x="RowNamedfWITHconta",by.y="RowNamesfdata",all.x=TRUE)
  DataToExport<-dfWITHconta[,c(2:MyNcol,(MyNcol+7):ncol(dfWITHconta))]
  
  colnames(DataToExport)[1:6]<-c("Protein.Group","Protein.Ids","Protein.Names","Genes","First.Protein.Description","Potential.contaminant")



#Export
if(MyStat==TRUE){
  if(MyPaired==TRUE){MyPaired<-".Paired"}else{MyPaired<-""}
  if(MySignifType=="P"){PQ<-"Pval"}else{PQ<-"Qval"}
  MySignifVal<-paste0(MySignifVal*100,"pc")
  if(MySignifFC==0){MySignifFC<-""}else{MySignifFC<-paste0("_FC",2^MySignifFC)}
}else{
  MyPaired<-""
  PQ<-""
  MySignifVal<-""
  MySignifFC<-""
  MyAlternative<-""
}

MyFileBase<-paste0(dirname(MyFile),"/",ProjectName,"_",MyAlternative,MyPaired,"_",PQ,MySignifVal,MySignifFC)

MyRDS<-list(edata=edata,fdata=fdata,phenodata=phenodata,Pepdata=Pepdata)
saveRDS(MyRDS, file = paste0(MyFileBase,".rds"))
write_xlsx(DataToExport,paste0(MyFileBase,".xlsx"))
write.table(DataToExport,paste0(MyFileBase,".txt"),row.names = FALSE,sep = "\t",quote=FALSE)

rm(MyNcol,DataToExport)

```
  
# 4. Contributions  
Samples preparation, LC MS-MS analysis :  
Proteins identification and quantification using DIANN, Data analysis :  
Expertise, strategy and Experimental design : 
